// @gmx-io/sdk/build/cjs/src/utils/__tests__/orderTransactions/decreaseOrders.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const viem_1 = require("viem");
const vitest_1 = require("vitest");
const chains_1 = require("../../../configs/chains");
const contracts_1 = require("../../../configs/contracts");
const markets_1 = require("../../../configs/markets");
const tokens_1 = require("../../../configs/tokens");
const orders_1 = require("../../../types/orders");
const numbers_1 = require("../../numbers");
const orderTransactions_1 = require("../../orderTransactions");
const uiFeeReceiver_1 = require("../../twap/uiFeeReceiver");
const mock_1 = require("../../../test/mock");
(0, vitest_1.beforeAll)(() => {
    vitest_1.vi.spyOn(Math, "random").mockReturnValue(0.5);
});
(0, vitest_1.describe)("Decrease Order Payloads", () => {
    const CHAIN_ID = chains_1.ARBITRUM;
    const RECEIVER = "0x1234567890123456789012345678901234567890";
    const UI_FEE_RECEIVER = "0x0987654321098765432109876543210987654321";
    const EXECUTION_GAS_LIMIT = 1000000n;
    const EXECUTION_FEE_AMOUNT = EXECUTION_GAS_LIMIT * mock_1.MOCK_GAS_PRICE;
    const REFERRAL_CODE = "0x999999cf1046e68e36E1aA2E0E07105eDDD1f08E";
    const USDC = (0, tokens_1.getTokenBySymbol)(CHAIN_ID, "USDC");
    const WETH = (0, tokens_1.getWrappedToken)(CHAIN_ID);
    const ETH_MARKET = markets_1.MARKETS[CHAIN_ID]["0x70d95587d40A2caf56bd97485aB3Eec10Bee6336"];
    const SLIPPAGE = 50; // 0.5%
    const ORDER_VAULT_ADDRESS = (0, contracts_1.getContract)(CHAIN_ID, "OrderVault");
    const commonParams = {
        chainId: CHAIN_ID,
        receiver: RECEIVER,
        uiFeeReceiver: UI_FEE_RECEIVER,
        executionFeeAmount: EXECUTION_FEE_AMOUNT,
        executionGasLimit: EXECUTION_GAS_LIMIT,
        referralCode: REFERRAL_CODE,
        validFromTime: 0n,
        autoCancel: false,
        marketAddress: ETH_MARKET.marketTokenAddress,
        indexTokenAddress: WETH.address,
        isLong: true,
        sizeDeltaUsd: (0, numbers_1.parseValue)("1000", numbers_1.USD_DECIMALS), // $1000
        sizeDeltaInTokens: (0, numbers_1.parseValue)("1", WETH.decimals), // 1 ETH
        collateralTokenAddress: WETH.address,
        collateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals), // 1 ETH
        swapPath: [ETH_MARKET.marketTokenAddress],
        orderType: orders_1.OrderType.MarketDecrease,
        allowedSlippage: SLIPPAGE,
        acceptablePrice: (0, numbers_1.parseValue)("1200", numbers_1.USD_DECIMALS), // $1200 base price
        triggerPrice: 0n,
        externalSwapQuote: undefined,
        minOutputUsd: 0n,
    };
    (0, vitest_1.describe)("buildDecreaseOrderPayload", () => {
        (0, vitest_1.it)("Market Decrease Long with Native Receive", () => {
            const params = {
                ...commonParams,
                isLong: true,
                decreasePositionSwapType: orders_1.DecreasePositionSwapType.SwapPnlTokenToCollateralToken,
                receiveTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
            };
            const result = (0, orderTransactions_1.buildDecreaseOrderPayload)(params);
            (0, vitest_1.expect)(result).toEqual({
                orderPayload: {
                    addresses: {
                        receiver: RECEIVER,
                        cancellationReceiver: viem_1.zeroAddress,
                        callbackContract: viem_1.zeroAddress,
                        uiFeeReceiver: UI_FEE_RECEIVER,
                        market: params.marketAddress,
                        initialCollateralToken: WETH.address,
                        swapPath: [ETH_MARKET.marketTokenAddress],
                    },
                    numbers: {
                        sizeDeltaUsd: (0, numbers_1.parseValue)("1000", numbers_1.USD_DECIMALS),
                        initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                        triggerPrice: 0n,
                        acceptablePrice: (0, numbers_1.parseValue)("1194", numbers_1.USD_DECIMALS - WETH.decimals), // $1200 - 0.5% slippage
                        executionFee: EXECUTION_FEE_AMOUNT,
                        callbackGasLimit: 0n,
                        minOutputAmount: 0n,
                        validFromTime: 0n,
                    },
                    orderType: orders_1.OrderType.MarketDecrease,
                    decreasePositionSwapType: orders_1.DecreasePositionSwapType.SwapPnlTokenToCollateralToken,
                    isLong: true,
                    shouldUnwrapNativeToken: true,
                    autoCancel: false,
                    referralCode: REFERRAL_CODE,
                    dataList: [],
                },
                params,
                tokenTransfersParams: {
                    isNativePayment: false,
                    isNativeReceive: true,
                    initialCollateralTokenAddress: WETH.address,
                    initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                    tokenTransfers: [
                        {
                            amount: EXECUTION_FEE_AMOUNT,
                            destination: ORDER_VAULT_ADDRESS,
                            tokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                        },
                    ],
                    payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                    payTokenAmount: 0n,
                    minOutputAmount: 0n,
                    swapPath: [ETH_MARKET.marketTokenAddress],
                    value: EXECUTION_FEE_AMOUNT,
                    externalCalls: undefined,
                },
            });
        });
        (0, vitest_1.it)("Market Decrease Short with Native Receive", () => {
            const params = {
                ...commonParams,
                isLong: false,
                decreasePositionSwapType: orders_1.DecreasePositionSwapType.SwapPnlTokenToCollateralToken,
                receiveTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
            };
            const result = (0, orderTransactions_1.buildDecreaseOrderPayload)(params);
            (0, vitest_1.expect)(result).toEqual({
                orderPayload: {
                    addresses: {
                        receiver: RECEIVER,
                        cancellationReceiver: viem_1.zeroAddress,
                        callbackContract: viem_1.zeroAddress,
                        uiFeeReceiver: UI_FEE_RECEIVER,
                        market: params.marketAddress,
                        initialCollateralToken: WETH.address,
                        swapPath: [ETH_MARKET.marketTokenAddress],
                    },
                    numbers: {
                        sizeDeltaUsd: (0, numbers_1.parseValue)("1000", numbers_1.USD_DECIMALS),
                        initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                        triggerPrice: 0n,
                        acceptablePrice: (0, numbers_1.parseValue)("1206", numbers_1.USD_DECIMALS - WETH.decimals), // $1200 + 0.5% slippage
                        executionFee: EXECUTION_FEE_AMOUNT,
                        callbackGasLimit: 0n,
                        minOutputAmount: 0n,
                        validFromTime: 0n,
                    },
                    orderType: orders_1.OrderType.MarketDecrease,
                    decreasePositionSwapType: orders_1.DecreasePositionSwapType.SwapPnlTokenToCollateralToken,
                    isLong: false,
                    shouldUnwrapNativeToken: true,
                    autoCancel: false,
                    referralCode: REFERRAL_CODE,
                    dataList: [],
                },
                params,
                tokenTransfersParams: {
                    isNativePayment: false,
                    isNativeReceive: true,
                    initialCollateralTokenAddress: WETH.address,
                    initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                    tokenTransfers: [
                        {
                            amount: EXECUTION_FEE_AMOUNT,
                            destination: ORDER_VAULT_ADDRESS,
                            tokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                        },
                    ],
                    payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                    payTokenAmount: 0n,
                    minOutputAmount: 0n,
                    swapPath: [ETH_MARKET.marketTokenAddress],
                    value: EXECUTION_FEE_AMOUNT,
                    externalCalls: undefined,
                },
            });
        });
        (0, vitest_1.it)("Market Decrease with ERC20 Receive", () => {
            const params = {
                ...commonParams,
                decreasePositionSwapType: orders_1.DecreasePositionSwapType.SwapPnlTokenToCollateralToken,
                receiveTokenAddress: USDC.address,
            };
            const result = (0, orderTransactions_1.buildDecreaseOrderPayload)(params);
            (0, vitest_1.expect)(result).toEqual({
                orderPayload: {
                    addresses: {
                        receiver: RECEIVER,
                        cancellationReceiver: viem_1.zeroAddress,
                        callbackContract: viem_1.zeroAddress,
                        uiFeeReceiver: UI_FEE_RECEIVER,
                        market: params.marketAddress,
                        initialCollateralToken: WETH.address,
                        swapPath: [ETH_MARKET.marketTokenAddress],
                    },
                    numbers: {
                        sizeDeltaUsd: (0, numbers_1.parseValue)("1000", numbers_1.USD_DECIMALS),
                        initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                        triggerPrice: 0n,
                        acceptablePrice: (0, numbers_1.parseValue)("1194", numbers_1.USD_DECIMALS - WETH.decimals),
                        executionFee: EXECUTION_FEE_AMOUNT,
                        callbackGasLimit: 0n,
                        minOutputAmount: 0n,
                        validFromTime: 0n,
                    },
                    orderType: orders_1.OrderType.MarketDecrease,
                    decreasePositionSwapType: orders_1.DecreasePositionSwapType.SwapPnlTokenToCollateralToken,
                    isLong: true,
                    shouldUnwrapNativeToken: false,
                    autoCancel: false,
                    referralCode: REFERRAL_CODE,
                    dataList: [],
                },
                params,
                tokenTransfersParams: {
                    isNativePayment: false,
                    isNativeReceive: false,
                    initialCollateralTokenAddress: WETH.address,
                    initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                    tokenTransfers: [
                        {
                            amount: EXECUTION_FEE_AMOUNT,
                            destination: ORDER_VAULT_ADDRESS,
                            tokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                        },
                    ],
                    payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                    payTokenAmount: 0n,
                    minOutputAmount: 0n,
                    swapPath: [ETH_MARKET.marketTokenAddress],
                    value: EXECUTION_FEE_AMOUNT,
                    externalCalls: undefined,
                },
            });
        });
        (0, vitest_1.it)("TP/SL Long", () => {
            const params = {
                ...commonParams,
                orderType: orders_1.OrderType.LimitDecrease,
                triggerPrice: (0, numbers_1.parseValue)("1200", numbers_1.USD_DECIMALS),
                decreasePositionSwapType: orders_1.DecreasePositionSwapType.SwapPnlTokenToCollateralToken,
                receiveTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
            };
            const result = (0, orderTransactions_1.buildDecreaseOrderPayload)(params);
            (0, vitest_1.expect)(result).toEqual({
                orderPayload: {
                    addresses: {
                        receiver: RECEIVER,
                        cancellationReceiver: viem_1.zeroAddress,
                        callbackContract: viem_1.zeroAddress,
                        uiFeeReceiver: UI_FEE_RECEIVER,
                        market: params.marketAddress,
                        initialCollateralToken: WETH.address,
                        swapPath: [ETH_MARKET.marketTokenAddress],
                    },
                    numbers: {
                        sizeDeltaUsd: (0, numbers_1.parseValue)("1000", numbers_1.USD_DECIMALS),
                        initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                        triggerPrice: (0, numbers_1.parseValue)("1200", numbers_1.USD_DECIMALS - WETH.decimals),
                        acceptablePrice: (0, numbers_1.parseValue)("1194", numbers_1.USD_DECIMALS - WETH.decimals),
                        executionFee: EXECUTION_FEE_AMOUNT,
                        callbackGasLimit: 0n,
                        minOutputAmount: 0n,
                        validFromTime: 0n,
                    },
                    orderType: orders_1.OrderType.LimitDecrease,
                    decreasePositionSwapType: orders_1.DecreasePositionSwapType.SwapPnlTokenToCollateralToken,
                    isLong: true,
                    shouldUnwrapNativeToken: true,
                    autoCancel: false,
                    referralCode: REFERRAL_CODE,
                    dataList: [],
                },
                params,
                tokenTransfersParams: {
                    isNativePayment: false,
                    isNativeReceive: true,
                    initialCollateralTokenAddress: WETH.address,
                    initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                    tokenTransfers: [
                        {
                            amount: EXECUTION_FEE_AMOUNT,
                            destination: ORDER_VAULT_ADDRESS,
                            tokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                        },
                    ],
                    payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                    payTokenAmount: 0n,
                    minOutputAmount: 0n,
                    swapPath: [ETH_MARKET.marketTokenAddress],
                    value: EXECUTION_FEE_AMOUNT,
                    externalCalls: undefined,
                },
            });
        });
        (0, vitest_1.it)("Auto cancel TP/SL Short", () => {
            const params = {
                ...commonParams,
                isLong: false,
                orderType: orders_1.OrderType.LimitDecrease,
                triggerPrice: (0, numbers_1.parseValue)("1200", numbers_1.USD_DECIMALS),
                decreasePositionSwapType: orders_1.DecreasePositionSwapType.SwapPnlTokenToCollateralToken,
                receiveTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                autoCancel: true,
            };
            const result = (0, orderTransactions_1.buildDecreaseOrderPayload)(params);
            (0, vitest_1.expect)(result).toEqual({
                orderPayload: {
                    addresses: {
                        receiver: RECEIVER,
                        cancellationReceiver: viem_1.zeroAddress,
                        callbackContract: viem_1.zeroAddress,
                        uiFeeReceiver: UI_FEE_RECEIVER,
                        market: params.marketAddress,
                        initialCollateralToken: WETH.address,
                        swapPath: [ETH_MARKET.marketTokenAddress],
                    },
                    numbers: {
                        sizeDeltaUsd: (0, numbers_1.parseValue)("1000", numbers_1.USD_DECIMALS),
                        initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                        triggerPrice: (0, numbers_1.parseValue)("1200", numbers_1.USD_DECIMALS - WETH.decimals),
                        acceptablePrice: (0, numbers_1.parseValue)("1206", numbers_1.USD_DECIMALS - WETH.decimals),
                        executionFee: EXECUTION_FEE_AMOUNT,
                        callbackGasLimit: 0n,
                        minOutputAmount: 0n,
                        validFromTime: 0n,
                    },
                    orderType: orders_1.OrderType.LimitDecrease,
                    decreasePositionSwapType: orders_1.DecreasePositionSwapType.SwapPnlTokenToCollateralToken,
                    isLong: false,
                    shouldUnwrapNativeToken: true,
                    autoCancel: true,
                    referralCode: REFERRAL_CODE,
                    dataList: [],
                },
                params,
                tokenTransfersParams: {
                    isNativePayment: false,
                    isNativeReceive: true,
                    initialCollateralTokenAddress: WETH.address,
                    initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                    tokenTransfers: [
                        {
                            amount: EXECUTION_FEE_AMOUNT,
                            destination: ORDER_VAULT_ADDRESS,
                            tokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                        },
                    ],
                    payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                    payTokenAmount: 0n,
                    minOutputAmount: 0n,
                    swapPath: [ETH_MARKET.marketTokenAddress],
                    value: EXECUTION_FEE_AMOUNT,
                    externalCalls: undefined,
                },
            });
        });
        (0, vitest_1.it)("TWAP Decrease Long", () => {
            const params = {
                ...commonParams,
                orderType: orders_1.OrderType.MarketDecrease,
                decreasePositionSwapType: orders_1.DecreasePositionSwapType.SwapPnlTokenToCollateralToken,
                receiveTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
            };
            const twapParams = {
                duration: {
                    hours: 1,
                    minutes: 0,
                },
                numberOfParts: 4,
            };
            const result = (0, orderTransactions_1.buildTwapOrdersPayloads)(params, twapParams);
            // Calculate time intervals for each part
            const totalDurationInSeconds = twapParams.duration.hours * 3600 + twapParams.duration.minutes * 60;
            const startTime = Math.ceil(Date.now() / 1000);
            const intervalInSeconds = totalDurationInSeconds / (twapParams.numberOfParts - 1);
            const expectedOrders = Array.from({ length: twapParams.numberOfParts }, (_, i) => {
                const validFromTime = BigInt(Math.floor(startTime + intervalInSeconds * i));
                const uiFeeReceiver = `0xff00000000000000000000000000000004800001`;
                const decoded = (0, uiFeeReceiver_1.decodeTwapUiFeeReceiver)(uiFeeReceiver);
                (0, vitest_1.expect)(decoded).toEqual({
                    twapId: "8000",
                    numberOfParts: twapParams.numberOfParts,
                    isExpress: false,
                });
                return {
                    orderPayload: {
                        addresses: {
                            receiver: RECEIVER,
                            cancellationReceiver: viem_1.zeroAddress,
                            callbackContract: viem_1.zeroAddress,
                            uiFeeReceiver,
                            market: params.marketAddress,
                            initialCollateralToken: WETH.address,
                            swapPath: [ETH_MARKET.marketTokenAddress],
                        },
                        numbers: {
                            sizeDeltaUsd: (0, numbers_1.parseValue)("250", numbers_1.USD_DECIMALS), // 1000/4
                            initialCollateralDeltaAmount: params.collateralDeltaAmount / 4n, // 1/4
                            triggerPrice: 0n,
                            acceptablePrice: 0n,
                            executionFee: EXECUTION_FEE_AMOUNT / 4n,
                            callbackGasLimit: 0n,
                            minOutputAmount: 0n,
                            validFromTime,
                        },
                        orderType: orders_1.OrderType.LimitDecrease,
                        decreasePositionSwapType: orders_1.DecreasePositionSwapType.SwapPnlTokenToCollateralToken,
                        isLong: true,
                        shouldUnwrapNativeToken: true,
                        autoCancel: false,
                        referralCode: REFERRAL_CODE,
                        dataList: [],
                    },
                    params: {
                        ...params,
                        acceptablePrice: 0n,
                        triggerPrice: 0n,
                        collateralDeltaAmount: params.collateralDeltaAmount / 4n,
                        executionFeeAmount: params.executionFeeAmount / 4n,
                        sizeDeltaUsd: params.sizeDeltaUsd / 4n,
                        sizeDeltaInTokens: params.sizeDeltaInTokens / 4n,
                        orderType: orders_1.OrderType.LimitDecrease,
                        allowedSlippage: 0,
                        uiFeeReceiver,
                        validFromTime,
                        referralCode: REFERRAL_CODE,
                    },
                    tokenTransfersParams: {
                        isNativePayment: false,
                        isNativeReceive: true,
                        initialCollateralTokenAddress: WETH.address,
                        initialCollateralDeltaAmount: params.collateralDeltaAmount / 4n,
                        tokenTransfers: [
                            {
                                amount: EXECUTION_FEE_AMOUNT / 4n,
                                destination: ORDER_VAULT_ADDRESS,
                                tokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                            },
                        ],
                        payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                        payTokenAmount: 0n,
                        minOutputAmount: 0n,
                        swapPath: [ETH_MARKET.marketTokenAddress],
                        value: EXECUTION_FEE_AMOUNT / 4n,
                        externalCalls: undefined,
                    },
                };
            });
            (0, vitest_1.expect)(result).toEqual(expectedOrders);
            (0, vitest_1.expect)(expectedOrders.every((co) => (0, orderTransactions_1.getIsTwapOrderPayload)(co.orderPayload))).toBe(true);
        });
        (0, vitest_1.it)("TWAP Decrease Short", () => {
            const params = {
                ...commonParams,
                isLong: false,
                orderType: orders_1.OrderType.MarketDecrease,
                decreasePositionSwapType: orders_1.DecreasePositionSwapType.SwapPnlTokenToCollateralToken,
                receiveTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
            };
            const twapParams = {
                duration: {
                    hours: 1,
                    minutes: 0,
                },
                numberOfParts: 4,
            };
            const result = (0, orderTransactions_1.buildTwapOrdersPayloads)(params, twapParams);
            // Calculate time intervals for each part
            const totalDurationInSeconds = twapParams.duration.hours * 3600 + twapParams.duration.minutes * 60;
            const startTime = Math.ceil(Date.now() / 1000);
            const intervalInSeconds = totalDurationInSeconds / (twapParams.numberOfParts - 1);
            const expectedOrders = Array.from({ length: twapParams.numberOfParts }, (_, i) => {
                const validFromTime = BigInt(Math.floor(startTime + intervalInSeconds * i));
                const uiFeeReceiver = `0xff00000000000000000000000000000004800001`;
                const decoded = (0, uiFeeReceiver_1.decodeTwapUiFeeReceiver)(uiFeeReceiver);
                (0, vitest_1.expect)(decoded).toEqual({
                    isExpress: false,
                    twapId: "8000",
                    numberOfParts: twapParams.numberOfParts,
                });
                return {
                    orderPayload: {
                        addresses: {
                            receiver: RECEIVER,
                            cancellationReceiver: viem_1.zeroAddress,
                            callbackContract: viem_1.zeroAddress,
                            uiFeeReceiver,
                            market: params.marketAddress,
                            initialCollateralToken: WETH.address,
                            swapPath: [ETH_MARKET.marketTokenAddress],
                        },
                        numbers: {
                            sizeDeltaUsd: (0, numbers_1.parseValue)("250", numbers_1.USD_DECIMALS), // 1000/4
                            initialCollateralDeltaAmount: params.collateralDeltaAmount / 4n, // 1/4
                            triggerPrice: numbers_1.MaxUint256,
                            acceptablePrice: numbers_1.MaxUint256,
                            executionFee: EXECUTION_FEE_AMOUNT / 4n,
                            callbackGasLimit: 0n,
                            minOutputAmount: 0n,
                            validFromTime,
                        },
                        orderType: orders_1.OrderType.LimitDecrease,
                        decreasePositionSwapType: orders_1.DecreasePositionSwapType.SwapPnlTokenToCollateralToken,
                        isLong: false,
                        shouldUnwrapNativeToken: true,
                        autoCancel: false,
                        referralCode: REFERRAL_CODE,
                        dataList: [],
                    },
                    params: {
                        ...params,
                        acceptablePrice: numbers_1.MaxUint256,
                        triggerPrice: numbers_1.MaxUint256,
                        collateralDeltaAmount: params.collateralDeltaAmount / 4n,
                        executionFeeAmount: params.executionFeeAmount / 4n,
                        sizeDeltaUsd: params.sizeDeltaUsd / 4n,
                        sizeDeltaInTokens: params.sizeDeltaInTokens / 4n,
                        orderType: orders_1.OrderType.LimitDecrease,
                        allowedSlippage: 0,
                        uiFeeReceiver,
                        validFromTime,
                    },
                    tokenTransfersParams: {
                        isNativePayment: false,
                        isNativeReceive: true,
                        initialCollateralTokenAddress: WETH.address,
                        initialCollateralDeltaAmount: params.collateralDeltaAmount / 4n,
                        tokenTransfers: [
                            {
                                amount: EXECUTION_FEE_AMOUNT / 4n,
                                destination: ORDER_VAULT_ADDRESS,
                                tokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                            },
                        ],
                        payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                        payTokenAmount: 0n,
                        minOutputAmount: 0n,
                        swapPath: [ETH_MARKET.marketTokenAddress],
                        value: EXECUTION_FEE_AMOUNT / 4n,
                        externalCalls: undefined,
                    },
                };
            });
            (0, vitest_1.expect)(result).toEqual(expectedOrders);
            (0, vitest_1.expect)(expectedOrders.every((co) => (0, orderTransactions_1.getIsTwapOrderPayload)(co.orderPayload))).toBe(true);
        });
        (0, vitest_1.it)("Market Decrease with Undefined Optional Parameters", () => {
            const params = {
                ...commonParams,
                orderType: orders_1.OrderType.MarketDecrease,
                decreasePositionSwapType: orders_1.DecreasePositionSwapType.SwapPnlTokenToCollateralToken,
                receiveTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                uiFeeReceiver: undefined,
                referralCode: undefined,
                validFromTime: undefined,
            };
            const result = (0, orderTransactions_1.buildDecreaseOrderPayload)(params);
            (0, vitest_1.expect)(result).toEqual({
                orderPayload: {
                    addresses: {
                        receiver: RECEIVER,
                        cancellationReceiver: viem_1.zeroAddress,
                        callbackContract: viem_1.zeroAddress,
                        uiFeeReceiver: viem_1.zeroAddress,
                        market: params.marketAddress,
                        initialCollateralToken: WETH.address,
                        swapPath: [ETH_MARKET.marketTokenAddress],
                    },
                    numbers: {
                        sizeDeltaUsd: (0, numbers_1.parseValue)("1000", numbers_1.USD_DECIMALS),
                        initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                        triggerPrice: 0n,
                        acceptablePrice: (0, numbers_1.parseValue)("1194", numbers_1.USD_DECIMALS - WETH.decimals),
                        executionFee: EXECUTION_FEE_AMOUNT,
                        callbackGasLimit: 0n,
                        minOutputAmount: 0n,
                        validFromTime: 0n,
                    },
                    orderType: orders_1.OrderType.MarketDecrease,
                    decreasePositionSwapType: orders_1.DecreasePositionSwapType.SwapPnlTokenToCollateralToken,
                    isLong: true,
                    shouldUnwrapNativeToken: true,
                    autoCancel: false,
                    referralCode: viem_1.zeroHash,
                    dataList: [],
                },
                params,
                tokenTransfersParams: {
                    isNativePayment: false,
                    isNativeReceive: true,
                    initialCollateralTokenAddress: WETH.address,
                    initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                    tokenTransfers: [
                        {
                            amount: EXECUTION_FEE_AMOUNT,
                            destination: ORDER_VAULT_ADDRESS,
                            tokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                        },
                    ],
                    payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                    payTokenAmount: 0n,
                    minOutputAmount: 0n,
                    swapPath: [ETH_MARKET.marketTokenAddress],
                    value: EXECUTION_FEE_AMOUNT,
                    externalCalls: undefined,
                },
            });
        });
    });
});
