// @gmx-io/sdk/build/cjs/src/utils/__tests__/orderTransactions/updateOrder.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const chains_1 = require("../../../configs/chains");
const tokens_1 = require("../../../configs/tokens");
const orders_1 = require("../../../types/orders");
const numbers_1 = require("../../numbers");
const orderTransactions_1 = require("../../orderTransactions");
const mock_1 = require("../../../test/mock");
(0, vitest_1.beforeAll)(() => {
    vitest_1.vi.spyOn(Math, "random").mockReturnValue(0.5);
});
(0, vitest_1.describe)("Update Order Payloads", () => {
    const CHAIN_ID = chains_1.ARBITRUM;
    const ORDER_KEY = "0x1234567890123456789012345678901234567890123456789012345678901234";
    const EXECUTION_GAS_LIMIT = 1000000n;
    const EXECUTION_FEE_AMOUNT = EXECUTION_GAS_LIMIT * mock_1.MOCK_GAS_PRICE;
    const USDC = (0, tokens_1.getTokenBySymbol)(CHAIN_ID, "USDC");
    const WETH = (0, tokens_1.getWrappedToken)(CHAIN_ID);
    (0, vitest_1.describe)("buildUpdateOrderPayload", () => {
        (0, vitest_1.it)("Update Limit Swap", () => {
            const params = {
                chainId: CHAIN_ID,
                orderKey: ORDER_KEY,
                orderType: orders_1.OrderType.LimitSwap,
                sizeDeltaUsd: 0n,
                triggerPrice: (0, numbers_1.parseValue)("1", 30),
                acceptablePrice: 0n,
                minOutputAmount: (0, numbers_1.parseValue)("0.8", USDC.decimals),
                autoCancel: false,
                executionFeeTopUp: 0n,
                indexTokenAddress: WETH.address,
                validFromTime: 0n,
            };
            const result = (0, orderTransactions_1.buildUpdateOrderPayload)(params);
            (0, vitest_1.expect)(result).toEqual({
                params,
                updatePayload: {
                    orderKey: ORDER_KEY,
                    sizeDeltaUsd: 0n,
                    triggerPrice: (0, numbers_1.parseValue)("1", 30),
                    acceptablePrice: 0n,
                    minOutputAmount: (0, numbers_1.parseValue)("0.8", USDC.decimals),
                    autoCancel: false,
                    validFromTime: 0n,
                    executionFeeTopUp: 0n,
                },
            });
        });
        (0, vitest_1.it)("Update Limit Increase with Execution Fee Top Up", () => {
            const params = {
                chainId: CHAIN_ID,
                orderKey: ORDER_KEY,
                orderType: orders_1.OrderType.LimitIncrease,
                sizeDeltaUsd: (0, numbers_1.parseValue)("1000", numbers_1.USD_DECIMALS), // $1000
                triggerPrice: (0, numbers_1.parseValue)("1200", numbers_1.USD_DECIMALS), // $1200
                acceptablePrice: (0, numbers_1.parseValue)("1194", numbers_1.USD_DECIMALS), // $1200 - 0.5% slippage
                minOutputAmount: 0n,
                autoCancel: false,
                executionFeeTopUp: EXECUTION_FEE_AMOUNT,
                indexTokenAddress: WETH.address,
                validFromTime: 0n,
            };
            const result = (0, orderTransactions_1.buildUpdateOrderPayload)(params);
            (0, vitest_1.expect)(result).toEqual({
                params,
                updatePayload: {
                    orderKey: ORDER_KEY,
                    sizeDeltaUsd: (0, numbers_1.parseValue)("1000", numbers_1.USD_DECIMALS),
                    triggerPrice: (0, numbers_1.parseValue)("1200", numbers_1.USD_DECIMALS - WETH.decimals), // Converted to contract price
                    acceptablePrice: (0, numbers_1.parseValue)("1194", numbers_1.USD_DECIMALS - WETH.decimals), // Converted to contract price
                    minOutputAmount: 0n,
                    autoCancel: false,
                    validFromTime: 0n,
                    executionFeeTopUp: EXECUTION_FEE_AMOUNT,
                },
            });
        });
    });
});
