// @gmx-io/sdk/build/cjs/src/utils/__tests__/parseError.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const errors_1 = require("../errors");
const transactionsErrors_1 = require("../errors/transactionsErrors");
(0, vitest_1.describe)("parseError", () => {
    (0, vitest_1.describe)("general errors", () => {
        (0, vitest_1.it)("should handle basic Error objects", () => {
            const error = new Error("Something went wrong");
            const result = (0, errors_1.parseError)(error);
            (0, vitest_1.expect)(result).toEqual(vitest_1.expect.objectContaining({
                errorMessage: "Something went wrong",
                errorName: "Error",
                errorGroup: "Something went wrong",
                errorStack: error.stack,
                errorStackHash: vitest_1.expect.any(String),
                errorDepth: 0,
            }));
        });
        (0, vitest_1.it)("should handle string errors", () => {
            const result = (0, errors_1.parseError)("API request failed");
            (0, vitest_1.expect)(result).toEqual(vitest_1.expect.objectContaining({
                errorMessage: "API request failed",
                errorGroup: "API request failed",
                errorStackGroup: "Unknown stack group",
                errorDepth: 0,
            }));
        });
        (0, vitest_1.it)("should handle nested errors with context", () => {
            const error = {
                message: "Order execution failed",
                errorContext: "simulation",
                info: {
                    error: {
                        message: "Price impact too high",
                        name: "ValidationError",
                    },
                },
            };
            const result = (0, errors_1.parseError)(error);
            (0, vitest_1.expect)(result).toEqual(vitest_1.expect.objectContaining({
                errorMessage: "Price impact too high",
                errorName: "ValidationError",
                errorContext: "simulation",
                errorDepth: 0,
            }));
        });
        (0, vitest_1.it)("should handle undefined errors", () => {
            const result = (0, errors_1.parseError)(undefined);
            (0, vitest_1.expect)(result).toEqual(vitest_1.expect.objectContaining({
                errorMessage: "undefined",
                errorGroup: "undefined",
                errorDepth: 0,
            }));
        });
        (0, vitest_1.describe)("error masking", () => {
            (0, vitest_1.it)("should mask URLs in error groups", () => {
                const error = new Error("Failed to fetch data from https://api.example.com:8080/v1/data?param=123");
                const result = (0, errors_1.parseError)(error);
                (0, vitest_1.expect)(result).toEqual(vitest_1.expect.objectContaining({
                    errorMessage: "Failed to fetch data from https://api.example.com:8080/v1/data?param=123",
                    errorGroup: "Failed to fetch data from https://api.example.com:",
                    errorDepth: 0,
                }));
            });
            (0, vitest_1.it)("should mask numbers in error groups", () => {
                const error = new Error("Transaction failed with gas 123456 at block 789012");
                const result = (0, errors_1.parseError)(error);
                (0, vitest_1.expect)(result).toEqual(vitest_1.expect.objectContaining({
                    errorMessage: "Transaction failed with gas 123456 at block 789012",
                    errorGroup: "Transaction failed with gas XXX at block XXX",
                    errorDepth: 0,
                }));
            });
            (0, vitest_1.it)("should mask both URLs and numbers in error groups", () => {
                const error = new Error("Failed  https://api.example.com/v1/tx/123456 with status 404 and error code E123");
                const result = (0, errors_1.parseError)(error);
                (0, vitest_1.expect)(result).toEqual(vitest_1.expect.objectContaining({
                    errorMessage: "Failed  https://api.example.com/v1/tx/123456 with status 404 and error code E123",
                    errorGroup: "Failed  https://api.example.com with status XXX an",
                    errorDepth: 0,
                }));
            });
            (0, vitest_1.it)("should mask URLs in stack traces", () => {
                const error = new Error("Processing failed");
                // Simulate a stack trace with URLs
                error.stack = `Error: Processing failed
            at processData (https://app.example.com/static/js/main.123456.js:12:34)
            at handleRequest (https://app.example.com/static/js/vendor.789012.js:56:78)`;
                const result = (0, errors_1.parseError)(error);
                (0, vitest_1.expect)(result?.errorStackGroup).toBe(`Error: Processing failed
            at processData (https://app.example.com)
            at handleRequest (https://app.example.com)`);
            });
        });
    });
    (0, vitest_1.describe)("transaction errors", () => {
        (0, vitest_1.it)("should handle ethers v6 user rejected error", () => {
            const error = {
                info: {
                    error: {
                        code: "ACTION_REJECTED",
                        message: "User denied transaction signature",
                    },
                },
            };
            const result = (0, errors_1.parseError)(error);
            (0, vitest_1.expect)(result).toEqual(vitest_1.expect.objectContaining({
                errorMessage: "User denied transaction signature",
                txErrorType: transactionsErrors_1.TxErrorType.UserDenied,
                isUserError: true,
                isUserRejectedError: true,
                errorGroup: "Txn Error: USER_DENIED",
            }));
        });
        (0, vitest_1.it)("should handle ethers v6 insufficient funds error", () => {
            const error = {
                info: {
                    error: {
                        message: "insufficient funds for gas",
                    },
                },
            };
            const result = (0, errors_1.parseError)(error);
            (0, vitest_1.expect)(result).toEqual(vitest_1.expect.objectContaining({
                errorMessage: "insufficient funds for gas",
                txErrorType: transactionsErrors_1.TxErrorType.NotEnoughFunds,
                isUserError: true,
                isUserRejectedError: false,
                errorGroup: "Txn Error: NOT_ENOUGH_FUNDS",
            }));
        });
        (0, vitest_1.it)("should handle RPC errors with code", () => {
            const error = {
                info: {
                    error: {
                        code: -32603,
                        message: "Internal JSON-RPC error",
                    },
                },
            };
            const result = (0, errors_1.parseError)(error);
            (0, vitest_1.expect)(result).toEqual(vitest_1.expect.objectContaining({
                errorMessage: "Internal JSON-RPC error",
                txErrorType: transactionsErrors_1.TxErrorType.RpcError,
                isUserError: false,
                isUserRejectedError: false,
                errorGroup: "Txn Error: RPC_ERROR",
            }));
        });
        (0, vitest_1.it)("should handle nested error in error.body", () => {
            const error = {
                info: {
                    error: {
                        message: JSON.stringify({
                            error: {
                                code: -32000,
                                message: "Invalid input parameters",
                            },
                        }),
                    },
                },
            };
            const result = (0, errors_1.parseError)(error);
            (0, vitest_1.expect)(result).toEqual(vitest_1.expect.objectContaining({
                errorMessage: JSON.stringify({
                    error: {
                        code: -32000,
                        message: "Invalid input parameters",
                    },
                }),
                errorGroup: '{"error":{"code":-XXX,"message":"Invalid input par',
                errorStackGroup: "Unknown stack group",
                errorDepth: 0,
                isUserError: false,
                isUserRejectedError: false,
            }));
        });
        (0, vitest_1.it)("should handle contract errors with data", () => {
            const error = {
                message: "execution reverted",
                data: "0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001a4f726465724e6f7446756c66696c6c61626c6541744c696d697400000000000000",
            };
            const result = (0, errors_1.parseError)(error);
            (0, vitest_1.expect)(result).toEqual(vitest_1.expect.objectContaining({
                errorMessage: "execution reverted",
                errorGroup: "execution reverted",
                errorStackGroup: "Unknown stack group",
                txErrorData: "0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001a4f726465724e6f7446756c66696c6c61626c6541744c696d697400000000000000",
                isUserError: false,
                isUserRejectedError: false,
            }));
        });
        (0, vitest_1.it)("should handle parent errors", () => {
            const error = {
                message: "Failed to execute transaction",
                parentError: {
                    message: "User denied transaction signature",
                    info: {
                        error: {
                            code: "ACTION_REJECTED",
                        },
                    },
                },
            };
            const result = (0, errors_1.parseError)(error);
            (0, vitest_1.expect)(result).toEqual(vitest_1.expect.objectContaining({
                errorMessage: "Failed to execute transaction",
                errorGroup: "Failed to execute transaction",
                errorStackGroup: "Unknown stack group",
                isUserError: false,
                isUserRejectedError: false,
                parentError: vitest_1.expect.objectContaining({
                    errorMessage: undefined,
                    errorGroup: "Unknown group",
                    errorStackGroup: "Unknown stack group",
                    errorDepth: 1,
                    isUserError: false,
                    isUserRejectedError: false,
                }),
            }));
        });
        (0, vitest_1.it)("should handle slippage errors", () => {
            const error = {
                message: "Router: mark price lower than limit",
            };
            const result = (0, errors_1.parseError)(error);
            (0, vitest_1.expect)(result).toEqual(vitest_1.expect.objectContaining({
                errorMessage: "Router: mark price lower than limit",
                txErrorType: transactionsErrors_1.TxErrorType.Slippage,
                isUserError: false,
                errorGroup: "Txn Error: SLIPPAGE",
            }));
        });
        (0, vitest_1.it)("should handle network change errors", () => {
            const error = {
                message: "network changed",
            };
            const result = (0, errors_1.parseError)(error);
            (0, vitest_1.expect)(result).toEqual(vitest_1.expect.objectContaining({
                errorMessage: "network changed",
                txErrorType: transactionsErrors_1.TxErrorType.NetworkChanged,
                isUserError: true,
                errorGroup: "Txn Error: NETWORK_CHANGED",
            }));
        });
        (0, vitest_1.it)("should handle additional validation info", () => {
            const error = {
                message: "Transaction failed",
                errorSource: "getCallStaticError",
                isAdditionalValidationPassed: false,
                additionalValidationType: "tryCallStatic",
            };
            const result = (0, errors_1.parseError)(error);
            (0, vitest_1.expect)(result).toEqual(vitest_1.expect.objectContaining({
                errorMessage: "Transaction failed",
                errorSource: "getCallStaticError",
                isAdditionalValidationPassed: false,
                additionalValidationType: "tryCallStatic",
            }));
        });
    });
});
