// @gmx-io/sdk/build/esm/src/modules/oracle.js
import fetch from "cross-fetch";
import { buildUrl } from "../utils/buildUrl";
export class Oracle {
    constructor(sdk) {
        this.sdk = sdk;
        this.url = sdk.config.oracleUrl;
    }
    getMarkets() {
        return fetch(buildUrl(this.url, "/markets"))
            .then((res) => res.json())
            .then((res) => {
            if (!res.markets || !res.markets.length) {
                throw new Error("Invalid markets response");
            }
            return res.markets;
        });
    }
    getTokens() {
        return fetch(buildUrl(this.url, "/tokens"))
            .then((res) => res.json())
            .then((res) => res.tokens.map(({ synthetic, ...rest }) => {
            return {
                ...rest,
                isSynthetic: synthetic,
            };
        }));
    }
    getTickers() {
        return fetch(buildUrl(this.url, "/prices/tickers"))
            .then((res) => res.json())
            .then((res) => {
            if (!res.length) {
                throw new Error("Invalid tickers response");
            }
            return res;
        });
    }
}
