// @gmx-io/sdk/build/cjs/src/modules/trades/trades.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchTradeActions = exports.Trades = void 0;
const merge_1 = __importDefault(require("lodash/merge"));
const tokens_1 = require("../../configs/tokens");
const orders_1 = require("../../types/orders");
const tradeHistory_1 = require("../../types/tradeHistory");
const graphqlFetcher_1 = __importDefault(require("../../utils/graphqlFetcher"));
const indexers_1 = require("../../utils/indexers");
const orders_2 = require("../../utils/orders");
const swapStats_1 = require("../../utils/swap/swapStats");
const tradeHistory_2 = require("../../utils/tradeHistory");
const base_1 = require("../base");
class Trades extends base_1.Module {
    async getTradeHistory(p) {
        const account = this.account;
        const { pageSize, forAllAccounts, fromTxTimestamp, toTxTimestamp, marketsDirectionsFilter, orderEventCombinations, marketsInfoData, pageIndex, tokensData, } = p;
        const data = await fetchTradeActions({
            sdk: this.sdk,
            pageIndex,
            pageSize,
            marketsDirectionsFilter,
            forAllAccounts,
            account,
            fromTxTimestamp,
            toTxTimestamp,
            orderEventCombinations,
            marketsInfoData,
            tokensData,
        });
        return data?.flat().filter(Boolean);
    }
}
exports.Trades = Trades;
async function fetchTradeActions({ sdk, pageIndex, pageSize, marketsDirectionsFilter = [], forAllAccounts, account, fromTxTimestamp, toTxTimestamp, orderEventCombinations, marketsInfoData, tokensData, }) {
    const endpoint = sdk.config.subsquidUrl;
    const chainId = sdk.chainId;
    if (!endpoint) {
        return [];
    }
    const offset = pageIndex * pageSize;
    const limit = pageSize;
    const nonSwapRelevantDefinedFiltersLowercased = marketsDirectionsFilter
        .filter((filter) => filter.direction !== "swap" && filter.marketAddress !== "any")
        .map((filter) => ({
        marketAddress: filter.marketAddress.toLowerCase(),
        direction: filter.direction,
        collateralAddress: filter.collateralAddress?.toLowerCase(),
    }));
    const hasNonSwapRelevantDefinedMarkets = nonSwapRelevantDefinedFiltersLowercased.length > 0;
    const pureDirectionFilters = marketsDirectionsFilter
        .filter((filter) => filter.direction !== "any" && filter.marketAddress === "any")
        .map((filter) => ({
        marketAddress: filter.marketAddress.toLowerCase(),
        direction: filter.direction,
    }));
    const hasPureDirectionFilters = pureDirectionFilters.length > 0;
    const swapRelevantDefinedMarketsLowercased = marketsDirectionsFilter
        .filter((filter) => (filter.direction === "any" || filter.direction === "swap") && filter.marketAddress !== "any")
        .map((filter) => filter.marketAddress.toLowerCase());
    const hasSwapRelevantDefinedMarkets = swapRelevantDefinedMarketsLowercased.length > 0;
    const filtersStr = (0, indexers_1.buildFiltersBody)({
        AND: [
            {
                account_eq: forAllAccounts ? undefined : account,
                timestamp_gte: fromTxTimestamp,
                timestamp_lte: toTxTimestamp,
            },
            {
                OR: !hasPureDirectionFilters
                    ? undefined
                    : pureDirectionFilters.map((filter) => filter.direction === "swap"
                        ? {
                            orderType_in: [orders_1.OrderType.LimitSwap, orders_1.OrderType.MarketSwap],
                        }
                        : {
                            isLong_eq: filter.direction === "long",
                            orderType_not_in: [orders_1.OrderType.LimitSwap, orders_1.OrderType.MarketSwap],
                        }),
            },
            {
                OR: [
                    // For non-swap orders
                    {
                        AND: !hasNonSwapRelevantDefinedMarkets
                            ? undefined
                            : [
                                {
                                    orderType_not_in: [orders_1.OrderType.LimitSwap, orders_1.OrderType.MarketSwap],
                                },
                                {
                                    OR: nonSwapRelevantDefinedFiltersLowercased.map((filter) => ({
                                        marketAddress_eq: filter.marketAddress === "any" ? undefined : filter.marketAddress,
                                        isLong_eq: filter.direction === "any" ? undefined : filter.direction === "long",
                                        // Collateral filtering is done outside of graphql on the client
                                    })),
                                },
                            ],
                    },
                    // For defined markets on swap orders
                    {
                        AND: !hasSwapRelevantDefinedMarkets
                            ? undefined
                            : [
                                {
                                    orderType_in: [orders_1.OrderType.LimitSwap, orders_1.OrderType.MarketSwap],
                                },
                                {
                                    OR: [
                                        // Source token is not in swap path so we add it to the or filter
                                        {
                                            marketAddress_in: swapRelevantDefinedMarketsLowercased,
                                        },
                                    ].concat(swapRelevantDefinedMarketsLowercased.map((marketAddress) => ({
                                        swapPath_contains: [marketAddress],
                                    })) || []),
                                },
                            ],
                    },
                ],
            },
            {
                OR: orderEventCombinations?.map((combination) => {
                    let sizeDeltaUsdCondition = {};
                    if (combination.orderType !== undefined &&
                        [orders_1.OrderType.MarketDecrease, orders_1.OrderType.MarketIncrease].includes(combination.orderType)) {
                        if (combination.isDepositOrWithdraw) {
                            sizeDeltaUsdCondition = { sizeDeltaUsd: 0 };
                        }
                        else {
                            sizeDeltaUsdCondition = { sizeDeltaUsd_not: 0 };
                        }
                    }
                    return (0, merge_1.default)({
                        eventName_eq: combination.eventName,
                        orderType_eq: combination.orderType,
                    }, sizeDeltaUsdCondition);
                }),
            },
            {
                // We do not show create liquidation orders in the trade history, thus we filter it out
                // ... && not (liquidation && orderCreated) === ... && (not liquidation || not orderCreated)
                OR: [{ orderType_not_eq: orders_1.OrderType.Liquidation }, { eventName_not_eq: tradeHistory_1.TradeActionType.OrderCreated }],
            },
        ],
    });
    const whereClause = `where: ${filtersStr}`;
    const query = `{
        tradeActions(
            offset: ${offset},
            limit: ${limit},
            orderBy: transaction_timestamp_DESC,
            ${whereClause}
        ) {
            id
            eventName

            account
            marketAddress
            swapPath
            initialCollateralTokenAddress

            initialCollateralDeltaAmount
            sizeDeltaUsd
            triggerPrice
            acceptablePrice
            executionPrice
            minOutputAmount
            executionAmountOut

            priceImpactUsd
            priceImpactDiffUsd
            positionFeeAmount
            borrowingFeeAmount
            fundingFeeAmount
            pnlUsd
            basePnlUsd

            collateralTokenPriceMax
            collateralTokenPriceMin

            indexTokenPriceMin
            indexTokenPriceMax

            orderType
            orderKey
            isLong
            shouldUnwrapNativeToken

            reason
            reasonBytes
            timestamp

            transaction {
                timestamp
                hash
            }
        }
      }`;
    const result = await (0, graphqlFetcher_1.default)(endpoint, query);
    const rawTradeActions = result?.tradeActions || [];
    if (!marketsInfoData || !tokensData) {
        return [];
    }
    const wrappedToken = (0, tokens_1.getWrappedToken)(chainId);
    const transformer = (0, tradeHistory_2.createRawTradeActionTransformer)(marketsInfoData, wrappedToken, tokensData);
    let tradeActions = rawTradeActions.map(transformer).filter(Boolean);
    const collateralFilterTree = {
        long: {},
        short: {},
    };
    let hasCollateralFilter = false;
    marketsDirectionsFilter.forEach((filter) => {
        if (filter.direction === "any" || filter.direction === "swap" || !filter.collateralAddress) {
            return;
        }
        if (!collateralFilterTree[filter.direction]) {
            collateralFilterTree[filter.direction] = {};
        }
        if (!collateralFilterTree[filter.direction][filter.marketAddress]) {
            collateralFilterTree[filter.direction][filter.marketAddress] = {};
        }
        hasCollateralFilter = true;
        collateralFilterTree[filter.direction][filter.marketAddress][filter.collateralAddress] = true;
    });
    // Filter out trade actions that do not match the collateral filter
    // We do this on the client side because the collateral filtering is too complex to be done in the graphql query
    if (hasCollateralFilter) {
        tradeActions = tradeActions.filter((tradeAction) => {
            // All necessary filters for swaps are already applied in the graphql query
            if ((0, orders_2.isSwapOrderType)(tradeAction.orderType)) {
                return true;
            }
            const positionTradeAction = tradeAction;
            let collateralMatch = true;
            const desiredCollateralAddresses = collateralFilterTree[positionTradeAction.isLong ? "long" : "short"]?.[positionTradeAction.marketAddress];
            if ((0, orders_2.isLimitOrderType)(tradeAction.orderType)) {
                const wrappedToken = (0, tokens_1.getWrappedToken)(chainId);
                if (!marketsInfoData) {
                    collateralMatch = true;
                }
                else {
                    const { outTokenAddress } = (0, swapStats_1.getSwapPathOutputAddresses)({
                        marketsInfoData,
                        initialCollateralAddress: positionTradeAction.initialCollateralTokenAddress,
                        isIncrease: (0, orders_2.isIncreaseOrderType)(tradeAction.orderType),
                        shouldUnwrapNativeToken: positionTradeAction.shouldUnwrapNativeToken,
                        swapPath: tradeAction.swapPath,
                        wrappedNativeTokenAddress: wrappedToken.address,
                    });
                    collateralMatch =
                        outTokenAddress !== undefined && Boolean(desiredCollateralAddresses?.[outTokenAddress]);
                }
            }
            else if ((0, orders_2.isTriggerDecreaseOrderType)(tradeAction.orderType)) {
                collateralMatch = Boolean(desiredCollateralAddresses?.[positionTradeAction.initialCollateralTokenAddress]);
            }
            return collateralMatch;
        });
    }
    return tradeActions;
}
exports.fetchTradeActions = fetchTradeActions;
