// @gmx-io/sdk/build/esm/src/modules/orders/transactions/updateOrderTxn.js
import { encodeFunctionData } from "viem";
import { abis } from "../../../abis";
import { getContract } from "../../../configs/contracts";
import { convertToContractPrice } from "../../../utils/tokens";
export function updateOrderTxn(sdk, p) {
    const { orderKey, sizeDeltaUsd, triggerPrice, acceptablePrice, minOutputAmount, executionFee, indexToken, autoCancel, } = p;
    const router = getContract(sdk.chainId, "ExchangeRouter");
    const encodedPayload = createUpdateEncodedPayload({
        sdk,
        orderKey,
        sizeDeltaUsd,
        executionFee,
        indexToken,
        acceptablePrice,
        triggerPrice,
        minOutputAmount,
        autoCancel,
    });
    return sdk.callContract(router, abis.ExchangeRouter, "multicall", [encodedPayload], {
        value: executionFee != undefined && executionFee > 0 ? executionFee : undefined,
    });
}
export function createUpdateEncodedPayload({ sdk, orderKey, sizeDeltaUsd, executionFee, indexToken, acceptablePrice, triggerPrice, autoCancel, minOutputAmount, }) {
    const orderVaultAddress = getContract(sdk.chainId, "OrderVault");
    const multicall = [];
    if (executionFee != undefined && executionFee > 0) {
        multicall.push({ method: "sendWnt", params: [orderVaultAddress, executionFee] });
    }
    multicall.push({
        method: "updateOrder",
        params: [
            orderKey,
            sizeDeltaUsd,
            acceptablePrice !== undefined ? convertToContractPrice(acceptablePrice, indexToken?.decimals || 0) : 0n,
            triggerPrice !== undefined ? convertToContractPrice(triggerPrice, indexToken?.decimals || 0) : 0n,
            minOutputAmount,
            0n,
            autoCancel,
        ],
    });
    return multicall.filter(Boolean).map((call) => encodeFunctionData({
        abi: abis.ExchangeRouter,
        functionName: call.method,
        args: call.params,
    }));
}
