// @gmx-io/sdk/build/cjs/src/utils/__tests__/orderTransactions/batchUtils.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const viem_1 = require("viem");
const vitest_1 = require("vitest");
const chains_1 = require("../../../configs/chains");
const tokens_1 = require("../../../configs/tokens");
const orders_1 = require("../../../types/orders");
const numbers_1 = require("../../numbers");
const orderTransactions_1 = require("../../orderTransactions");
const mock_1 = require("../../../test/mock");
// Common tokens and addresses
const CHAIN_ID = chains_1.ARBITRUM;
const ACCOUNT = "0x1234567890123456789012345678901234567890";
const WETH = (0, tokens_1.getWrappedToken)(CHAIN_ID);
const USDC = (0, tokens_1.getTokenBySymbol)(CHAIN_ID, "USDC");
const USDT = (0, tokens_1.getTokenBySymbol)(CHAIN_ID, "USDT");
// Common MarketIncrease params
const commonMarketIncreaseParams = {
    chainId: CHAIN_ID,
    receiver: ACCOUNT,
    executionGasLimit: 0n,
    payTokenAddress: WETH.address,
    payTokenAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
    marketAddress: "0x1111111111111111111111111111111111111111",
    indexTokenAddress: WETH.address,
    isLong: true,
    sizeDeltaUsd: (0, numbers_1.parseValue)("1000", numbers_1.USD_DECIMALS),
    sizeDeltaInTokens: (0, numbers_1.parseValue)("1", WETH.decimals),
    acceptablePrice: (0, numbers_1.parseValue)("1200", numbers_1.USD_DECIMALS),
    collateralTokenAddress: WETH.address,
    collateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
    swapPath: [WETH.address],
    externalSwapQuote: undefined,
    triggerPrice: undefined,
    referralCode: viem_1.zeroHash,
    autoCancel: false,
    allowedSlippage: 100,
    executionFeeAmount: (0, numbers_1.parseValue)("0.1", WETH.decimals),
    validFromTime: 0n,
    orderType: orders_1.OrderType.MarketIncrease,
    uiFeeReceiver: viem_1.zeroAddress,
};
// Helper to build a batch with multiple increase orders
function buildMultiIncreaseBatch(paramsList) {
    return {
        createOrderParams: paramsList.map((p) => (0, orderTransactions_1.buildIncreaseOrderPayload)(p)),
        updateOrderParams: [],
        cancelOrderParams: [],
    };
}
(0, vitest_1.describe)("Batch Utils", () => {
    (0, vitest_1.describe)("getIsEmptyBatch", () => {
        (0, vitest_1.it)("undefined batch", () => {
            (0, vitest_1.expect)((0, orderTransactions_1.getIsEmptyBatch)(undefined)).toBe(true);
        });
        (0, vitest_1.it)("zero actions batch", () => {
            (0, vitest_1.expect)((0, orderTransactions_1.getIsEmptyBatch)({ createOrderParams: [], updateOrderParams: [], cancelOrderParams: [] })).toBe(true);
        });
        (0, vitest_1.it)("multiple orders batch with one empty order", () => {
            const emptyOrder = (0, orderTransactions_1.buildIncreaseOrderPayload)({
                ...commonMarketIncreaseParams,
                payTokenAmount: 0n,
                sizeDeltaUsd: 0n,
                sizeDeltaInTokens: 0n,
                collateralDeltaAmount: 0n,
            });
            const nonEmptyOrder = (0, orderTransactions_1.buildIncreaseOrderPayload)(commonMarketIncreaseParams);
            const batch = {
                createOrderParams: [emptyOrder, nonEmptyOrder],
                updateOrderParams: [],
                cancelOrderParams: [],
            };
            (0, vitest_1.expect)((0, orderTransactions_1.getIsEmptyBatch)(batch)).toBe(true);
        });
        (0, vitest_1.it)("multiple non-empty orders batch", () => {
            const batch = buildMultiIncreaseBatch([commonMarketIncreaseParams, commonMarketIncreaseParams]);
            (0, vitest_1.expect)((0, orderTransactions_1.getIsEmptyBatch)(batch)).toBe(false);
        });
    });
    (0, vitest_1.describe)("getBatchIsNativePayment", () => {
        (0, vitest_1.it)("returns true if any order has native payment", () => {
            const nativeParams = { ...commonMarketIncreaseParams, payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS };
            const batch = buildMultiIncreaseBatch([nativeParams]);
            (0, vitest_1.expect)((0, orderTransactions_1.getBatchIsNativePayment)(batch)).toBe(true);
        });
        (0, vitest_1.it)("returns false if no orders have native payment", () => {
            const batch = buildMultiIncreaseBatch([commonMarketIncreaseParams]);
            (0, vitest_1.expect)((0, orderTransactions_1.getBatchIsNativePayment)(batch)).toBe(false);
        });
    });
    (0, vitest_1.describe)("getIsInvalidBatchReceiver", () => {
        (0, vitest_1.it)("returns true if orders have different receivers", () => {
            const params1 = { ...commonMarketIncreaseParams, receiver: ACCOUNT };
            const params2 = {
                ...commonMarketIncreaseParams,
                receiver: "0x9999999999999999999999999999999999999999",
            };
            const batch = buildMultiIncreaseBatch([params1, params2]);
            (0, vitest_1.expect)((0, orderTransactions_1.getIsInvalidBatchReceiver)(batch, ACCOUNT)).toBe(true);
        });
        (0, vitest_1.it)("returns false if all orders have the same receiver", () => {
            const batch = buildMultiIncreaseBatch([commonMarketIncreaseParams, commonMarketIncreaseParams]);
            (0, vitest_1.expect)((0, orderTransactions_1.getIsInvalidBatchReceiver)(batch, ACCOUNT)).toBe(false);
        });
    });
    (0, vitest_1.describe)("getBatchExternalCalls", () => {
        (0, vitest_1.it)("combines external calls from multiple orders", () => {
            const params1 = {
                ...commonMarketIncreaseParams,
                payTokenAddress: WETH.address,
                payTokenAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                externalSwapQuote: (0, mock_1.mockExternalSwap)({
                    inToken: WETH,
                    outToken: USDC,
                    to: "0x6352a56caadC4F1E25CD6c75970Fa768A3304e64",
                    data: "0x1",
                    amountIn: (0, numbers_1.parseValue)("1", WETH.decimals),
                    amountOut: (0, numbers_1.parseValue)("1000", USDC.decimals),
                    priceIn: (0, numbers_1.expandDecimals)(1000, numbers_1.USD_DECIMALS),
                    priceOut: (0, numbers_1.expandDecimals)(1, numbers_1.USD_DECIMALS),
                }),
            };
            const params2 = {
                ...commonMarketIncreaseParams,
                payTokenAddress: USDC.address,
                payTokenAmount: (0, numbers_1.parseValue)("1000", USDC.decimals),
                externalSwapQuote: (0, mock_1.mockExternalSwap)({
                    inToken: USDC,
                    outToken: WETH,
                    to: "0x6352a56caadC4F1E25CD6c75970Fa768A3304e64",
                    data: "0x2",
                    amountIn: (0, numbers_1.parseValue)("1000", USDC.decimals),
                    amountOut: (0, numbers_1.parseValue)("0.5", WETH.decimals),
                    priceIn: (0, numbers_1.expandDecimals)(1, numbers_1.USD_DECIMALS),
                    priceOut: (0, numbers_1.expandDecimals)(2000, numbers_1.USD_DECIMALS),
                }),
            };
            const batch = buildMultiIncreaseBatch([params1, params2]);
            const result = (0, orderTransactions_1.getBatchExternalCalls)(batch);
            (0, vitest_1.expect)(result).toEqual({
                sendTokens: [WETH.address, USDC.address],
                sendAmounts: [(0, numbers_1.parseValue)("1", WETH.decimals), (0, numbers_1.parseValue)("1000", USDC.decimals)],
                externalCallTargets: [
                    "0x6352a56caadC4F1E25CD6c75970Fa768A3304e64",
                    "0x6352a56caadC4F1E25CD6c75970Fa768A3304e64",
                ],
                externalCallDataList: ["0x1", "0x2"],
                refundTokens: [WETH.address, USDC.address],
                refundReceivers: [ACCOUNT, ACCOUNT],
            });
        });
    });
    (0, vitest_1.describe)("getBatchTotalPayCollateralAmount", () => {
        (0, vitest_1.it)("sums pay amounts across orders", () => {
            const params1 = {
                ...commonMarketIncreaseParams,
                externalSwapQuote: undefined,
                payTokenAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
            };
            const params2 = {
                ...commonMarketIncreaseParams,
                externalSwapQuote: undefined,
                payTokenAmount: (0, numbers_1.parseValue)("2", WETH.decimals),
            };
            const params3 = {
                ...commonMarketIncreaseParams,
                externalSwapQuote: undefined,
                payTokenAddress: USDC.address,
                payTokenAmount: (0, numbers_1.parseValue)("3", USDC.decimals),
            };
            const batch = buildMultiIncreaseBatch([params1, params2, params3]);
            const result = (0, orderTransactions_1.getBatchTotalPayCollateralAmount)(batch);
            (0, vitest_1.expect)(result).toEqual({
                [WETH.address]: (0, numbers_1.parseValue)("3", WETH.decimals),
                [USDC.address]: (0, numbers_1.parseValue)("3", USDC.decimals),
            });
        });
    });
    (0, vitest_1.describe)("getBatchTotalExecutionFee", () => {
        (0, vitest_1.it)("calculates total execution fee including top-ups", () => {
            const params5 = {
                ...commonMarketIncreaseParams,
                externalSwapQuote: undefined,
                executionFeeAmount: (0, numbers_1.parseValue)("0.1", WETH.decimals),
            };
            const batch = buildMultiIncreaseBatch([params5]);
            batch.updateOrderParams = [{ updatePayload: { executionFeeTopUp: (0, numbers_1.parseValue)("0.05", WETH.decimals) } }];
            const tokensData = {
                [WETH.address]: {
                    ...WETH,
                    prices: {
                        minPrice: (0, numbers_1.expandDecimals)(2000, numbers_1.USD_DECIMALS),
                        maxPrice: (0, numbers_1.expandDecimals)(2000, numbers_1.USD_DECIMALS),
                    },
                },
            };
            const result = (0, orderTransactions_1.getBatchTotalExecutionFee)({ batchParams: batch, tokensData, chainId: CHAIN_ID });
            (0, vitest_1.expect)(result).toEqual({
                feeTokenAmount: (0, numbers_1.parseValue)("0.15", WETH.decimals),
                gasLimit: 0n,
                feeUsd: (0, numbers_1.expandDecimals)(300, numbers_1.USD_DECIMALS),
                feeToken: tokensData[WETH.address],
                isFeeHigh: true,
                isFeeVeryHigh: true,
            });
        });
    });
    (0, vitest_1.describe)("getBatchExternalSwapGasLimit", () => {
        (0, vitest_1.it)("sums gas limits from external swap quotes", () => {
            const quote1 = (0, mock_1.mockExternalSwap)({
                inToken: WETH,
                outToken: USDC,
                amountIn: (0, numbers_1.parseValue)("1", WETH.decimals),
                amountOut: (0, numbers_1.parseValue)("1000", USDC.decimals),
                priceIn: (0, numbers_1.expandDecimals)(1000, numbers_1.USD_DECIMALS),
                priceOut: (0, numbers_1.expandDecimals)(1, numbers_1.USD_DECIMALS),
            });
            const quote2 = (0, mock_1.mockExternalSwap)({
                inToken: USDC,
                outToken: USDT,
                amountIn: (0, numbers_1.parseValue)("1000", USDC.decimals),
                amountOut: (0, numbers_1.parseValue)("1000", USDT.decimals),
                priceIn: (0, numbers_1.expandDecimals)(1, numbers_1.USD_DECIMALS),
                priceOut: (0, numbers_1.expandDecimals)(1, numbers_1.USD_DECIMALS),
            });
            const params6 = {
                ...commonMarketIncreaseParams,
                externalSwapQuote: quote1,
            };
            const params7 = {
                ...commonMarketIncreaseParams,
                externalSwapQuote: quote2,
                payTokenAddress: USDC.address,
            };
            const batch = buildMultiIncreaseBatch([params6, params7]);
            const result = (0, orderTransactions_1.getBatchExternalSwapGasLimit)(batch);
            (0, vitest_1.expect)(result).toBe(quote1.txnData.estimatedGas + quote2.txnData.estimatedGas);
        });
    });
});
