// @gmx-io/sdk/build/cjs/src/utils/__tests__/orderTransactions/externalCalls.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const viem_1 = require("viem");
const vitest_1 = require("vitest");
const chains_1 = require("../../../configs/chains");
const tokens_1 = require("../../../configs/tokens");
const numbers_1 = require("../../numbers");
const orderTransactions_1 = require("../../orderTransactions");
const mock_1 = require("../../../test/mock");
(0, vitest_1.describe)("External Calls", () => {
    const CHAIN_ID = chains_1.ARBITRUM;
    const ACCOUNT = "0x1234567890123456789012345678901234567890";
    const WETH = (0, tokens_1.getWrappedToken)(CHAIN_ID);
    const USDC = (0, tokens_1.getTokenBySymbol)(CHAIN_ID, "USDC");
    const USDT = (0, tokens_1.getTokenBySymbol)(CHAIN_ID, "USDT");
    const TARGET_ADDRESS = "0x1111111111111111111111111111111111111111";
    const TARGET_ADDRESS_2 = "0x2222222222222222222222222222222222222222";
    (0, vitest_1.describe)("combineExternalCalls", () => {
        (0, vitest_1.it)("combines multiple external calls with repeated tokens", () => {
            const externalCalls = [
                {
                    sendTokens: [WETH.address, USDC.address],
                    sendAmounts: [(0, numbers_1.parseValue)("1", WETH.decimals), (0, numbers_1.parseValue)("1000", USDC.decimals)],
                    externalCallTargets: [TARGET_ADDRESS],
                    externalCallDataList: ["0x1111"],
                    refundTokens: [WETH.address, USDC.address],
                    refundReceivers: [ACCOUNT, ACCOUNT],
                },
                {
                    sendTokens: [WETH.address, USDT.address],
                    sendAmounts: [(0, numbers_1.parseValue)("2", WETH.decimals), (0, numbers_1.parseValue)("2000", USDT.decimals)],
                    externalCallTargets: [TARGET_ADDRESS],
                    externalCallDataList: ["0x2222"],
                    refundTokens: [USDT.address, WETH.address],
                    refundReceivers: [ACCOUNT, ACCOUNT],
                },
                {
                    sendTokens: [WETH.address],
                    sendAmounts: [(0, numbers_1.parseValue)("3", WETH.decimals)],
                    externalCallTargets: [TARGET_ADDRESS_2],
                    externalCallDataList: ["0x3333"],
                    refundTokens: [WETH.address],
                    refundReceivers: [ACCOUNT],
                },
            ];
            const result = (0, orderTransactions_1.combineExternalCalls)(externalCalls);
            (0, vitest_1.expect)(result).toEqual({
                sendTokens: [WETH.address, USDC.address, USDT.address],
                sendAmounts: [
                    (0, numbers_1.parseValue)("6", WETH.decimals), // 1 + 2 + 3 WETH
                    (0, numbers_1.parseValue)("1000", USDC.decimals), // 1000 USDC
                    (0, numbers_1.parseValue)("2000", USDT.decimals),
                ],
                externalCallTargets: [TARGET_ADDRESS, TARGET_ADDRESS, TARGET_ADDRESS_2],
                externalCallDataList: ["0x1111", "0x2222", "0x3333"],
                refundTokens: [WETH.address, USDC.address, USDT.address],
                refundReceivers: [ACCOUNT, ACCOUNT, ACCOUNT],
            });
        });
    });
    (0, vitest_1.describe)("getExternalCallsPayload", () => {
        (0, vitest_1.it)("creates payload without spender approval", () => {
            const quote = (0, mock_1.mockExternalSwap)({
                inToken: WETH,
                outToken: USDC,
                amountIn: (0, numbers_1.parseValue)("1", WETH.decimals),
                amountOut: (0, numbers_1.parseValue)("1000", USDC.decimals),
                priceIn: (0, numbers_1.expandDecimals)(1000, numbers_1.USD_DECIMALS),
                priceOut: (0, numbers_1.expandDecimals)(1, numbers_1.USD_DECIMALS),
                data: "0x1111",
                to: "0x1111111111111111111111111111111111111111",
            });
            const result = (0, orderTransactions_1.getExternalCallsPayload)({
                chainId: CHAIN_ID,
                account: ACCOUNT,
                quote,
            });
            (0, vitest_1.expect)(result).toEqual({
                sendTokens: [WETH.address],
                sendAmounts: [(0, numbers_1.parseValue)("1", WETH.decimals)],
                externalCallTargets: [quote.txnData.to],
                externalCallDataList: [quote.txnData.data],
                refundTokens: [WETH.address, USDC.address],
                refundReceivers: [ACCOUNT, ACCOUNT],
            });
        });
        (0, vitest_1.it)("creates payload with spender approval", () => {
            const quote = (0, mock_1.mockExternalSwap)({
                inToken: WETH,
                outToken: USDC,
                amountIn: (0, numbers_1.parseValue)("1", WETH.decimals),
                amountOut: (0, numbers_1.parseValue)("1000", USDC.decimals),
                priceIn: (0, numbers_1.expandDecimals)(1000, numbers_1.USD_DECIMALS),
                priceOut: (0, numbers_1.expandDecimals)(1, numbers_1.USD_DECIMALS),
                data: "0x1111",
                to: "0x1111111111111111111111111111111111111111",
            });
            const result = (0, orderTransactions_1.getExternalCallsPayload)({
                chainId: CHAIN_ID,
                account: ACCOUNT,
                quote: { ...quote, needSpenderApproval: true },
            });
            const expectedApproveData = (0, viem_1.encodeFunctionData)({
                abi: viem_1.erc20Abi,
                functionName: "approve",
                args: [quote.txnData.to, numbers_1.MaxUint256],
            });
            (0, vitest_1.expect)(result).toEqual({
                sendTokens: [WETH.address],
                sendAmounts: [(0, numbers_1.parseValue)("1", WETH.decimals)],
                externalCallTargets: [WETH.address, quote.txnData.to],
                externalCallDataList: [expectedApproveData, quote.txnData.data],
                refundTokens: [WETH.address, USDC.address],
                refundReceivers: [ACCOUNT, ACCOUNT],
            });
        });
    });
});
