// @gmx-io/sdk/build/cjs/src/utils/__tests__/orderTransactions/increaseOrders.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const viem_1 = require("viem");
const vitest_1 = require("vitest");
const chains_1 = require("../../../configs/chains");
const contracts_1 = require("../../../configs/contracts");
const markets_1 = require("../../../configs/markets");
const tokens_1 = require("../../../configs/tokens");
const orders_1 = require("../../../types/orders");
const numbers_1 = require("../../numbers");
const orderTransactions_1 = require("../../orderTransactions");
const uiFeeReceiver_1 = require("../../twap/uiFeeReceiver");
const mock_1 = require("../../../test/mock");
(0, vitest_1.beforeAll)(() => {
    vitest_1.vi.spyOn(Math, "random").mockReturnValue(0.5);
});
(0, vitest_1.describe)("Increase Order Payloads", () => {
    const CHAIN_ID = chains_1.ARBITRUM;
    const RECEIVER = "0x1234567890123456789012345678901234567890";
    const UI_FEE_RECEIVER = "0x0987654321098765432109876543210987654321";
    const EXECUTION_GAS_LIMIT = 1000000n;
    const EXECUTION_FEE_AMOUNT = EXECUTION_GAS_LIMIT * mock_1.MOCK_GAS_PRICE;
    const REFERRAL_CODE = "0xf2742351cc0eca941ff90bf489789ee6169cbeacfdd38eba60012218fac1b7e5";
    const USDC = (0, tokens_1.getTokenBySymbol)(CHAIN_ID, "USDC");
    const WETH = (0, tokens_1.getWrappedToken)(CHAIN_ID);
    const ETH_MARKET = markets_1.MARKETS[CHAIN_ID]["0x70d95587d40A2caf56bd97485aB3Eec10Bee6336"];
    const SLIPPAGE = 50; // 0.5%
    const ORDER_VAULT_ADDRESS = (0, contracts_1.getContract)(CHAIN_ID, "OrderVault");
    const EXTERNAL_HANDLER_ADDRESS = (0, contracts_1.getContract)(CHAIN_ID, "ExternalHandler");
    const commonParams = {
        chainId: CHAIN_ID,
        receiver: RECEIVER,
        uiFeeReceiver: UI_FEE_RECEIVER,
        executionFeeAmount: EXECUTION_FEE_AMOUNT,
        executionGasLimit: EXECUTION_GAS_LIMIT,
        referralCode: REFERRAL_CODE,
        validFromTime: 0n,
        autoCancel: false,
        marketAddress: ETH_MARKET.marketTokenAddress,
        indexTokenAddress: WETH.address,
        isLong: true,
        sizeDeltaUsd: (0, numbers_1.parseValue)("1000", numbers_1.USD_DECIMALS), // $1000
        sizeDeltaInTokens: (0, numbers_1.parseValue)("1", WETH.decimals), // 1 ETH
        collateralTokenAddress: WETH.address,
        collateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals), // 1 ETH
        swapPath: [ETH_MARKET.marketTokenAddress],
        orderType: orders_1.OrderType.MarketIncrease,
        allowedSlippage: SLIPPAGE,
        acceptablePrice: (0, numbers_1.parseValue)("1200", numbers_1.USD_DECIMALS), // $1200 base price
        triggerPrice: 0n,
        externalSwapQuote: undefined,
    };
    (0, vitest_1.describe)("buildIncreaseOrderPayload", () => {
        (0, vitest_1.it)("Market Increase Long Pay with Native Token", () => {
            const params = {
                ...commonParams,
                isLong: true,
                payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                payTokenAmount: (0, numbers_1.parseValue)("1", WETH.decimals), // 1 ETH
            };
            const result = (0, orderTransactions_1.buildIncreaseOrderPayload)(params);
            (0, vitest_1.expect)(result).toEqual({
                orderPayload: {
                    addresses: {
                        receiver: RECEIVER,
                        cancellationReceiver: viem_1.zeroAddress,
                        callbackContract: viem_1.zeroAddress,
                        uiFeeReceiver: UI_FEE_RECEIVER,
                        market: params.marketAddress,
                        initialCollateralToken: WETH.address,
                        swapPath: [ETH_MARKET.marketTokenAddress],
                    },
                    numbers: {
                        sizeDeltaUsd: (0, numbers_1.parseValue)("1000", numbers_1.USD_DECIMALS),
                        initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                        triggerPrice: 0n,
                        acceptablePrice: (0, numbers_1.parseValue)("1206", numbers_1.USD_DECIMALS - WETH.decimals), // $1200 + 0.5% slippage
                        executionFee: EXECUTION_FEE_AMOUNT,
                        callbackGasLimit: 0n,
                        minOutputAmount: 0n,
                        validFromTime: 0n,
                    },
                    orderType: orders_1.OrderType.MarketIncrease,
                    decreasePositionSwapType: orders_1.DecreasePositionSwapType.NoSwap,
                    isLong: true,
                    shouldUnwrapNativeToken: true,
                    autoCancel: false,
                    referralCode: REFERRAL_CODE,
                    dataList: [],
                },
                params,
                tokenTransfersParams: {
                    isNativePayment: true,
                    isNativeReceive: false,
                    initialCollateralTokenAddress: WETH.address,
                    initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                    tokenTransfers: [
                        {
                            amount: EXECUTION_FEE_AMOUNT + (0, numbers_1.parseValue)("1", WETH.decimals),
                            destination: ORDER_VAULT_ADDRESS,
                            tokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                        },
                    ],
                    payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                    payTokenAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                    minOutputAmount: 0n,
                    swapPath: [ETH_MARKET.marketTokenAddress],
                    value: EXECUTION_FEE_AMOUNT + (0, numbers_1.parseValue)("1", WETH.decimals),
                    externalCalls: undefined,
                },
            });
        });
        (0, vitest_1.it)("Market Increase Short Pay with Native Token", () => {
            const params = {
                ...commonParams,
                isLong: false,
                payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                payTokenAmount: (0, numbers_1.parseValue)("1", WETH.decimals), // 1 ETH
            };
            const result = (0, orderTransactions_1.buildIncreaseOrderPayload)(params);
            (0, vitest_1.expect)(result).toEqual({
                orderPayload: {
                    addresses: {
                        receiver: RECEIVER,
                        cancellationReceiver: viem_1.zeroAddress,
                        callbackContract: viem_1.zeroAddress,
                        uiFeeReceiver: UI_FEE_RECEIVER,
                        market: params.marketAddress,
                        initialCollateralToken: WETH.address,
                        swapPath: [ETH_MARKET.marketTokenAddress],
                    },
                    numbers: {
                        sizeDeltaUsd: (0, numbers_1.parseValue)("1000", numbers_1.USD_DECIMALS),
                        initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                        triggerPrice: 0n,
                        acceptablePrice: (0, numbers_1.parseValue)("1194", numbers_1.USD_DECIMALS - WETH.decimals), // $1200 - 0.5% slippage
                        executionFee: EXECUTION_FEE_AMOUNT,
                        callbackGasLimit: 0n,
                        minOutputAmount: 0n,
                        validFromTime: 0n,
                    },
                    orderType: orders_1.OrderType.MarketIncrease,
                    decreasePositionSwapType: orders_1.DecreasePositionSwapType.NoSwap,
                    isLong: false,
                    shouldUnwrapNativeToken: true,
                    autoCancel: false,
                    referralCode: REFERRAL_CODE,
                    dataList: [],
                },
                params,
                tokenTransfersParams: {
                    isNativePayment: true,
                    isNativeReceive: false,
                    initialCollateralTokenAddress: WETH.address,
                    initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                    tokenTransfers: [
                        {
                            amount: EXECUTION_FEE_AMOUNT + (0, numbers_1.parseValue)("1", WETH.decimals),
                            destination: ORDER_VAULT_ADDRESS,
                            tokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                        },
                    ],
                    payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                    payTokenAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                    minOutputAmount: 0n,
                    swapPath: [ETH_MARKET.marketTokenAddress],
                    value: EXECUTION_FEE_AMOUNT + (0, numbers_1.parseValue)("1", WETH.decimals),
                    externalCalls: undefined,
                },
            });
        });
        (0, vitest_1.it)("Market Increase Pay with ERC20", () => {
            const params = {
                ...commonParams,
                payTokenAddress: USDC.address,
                payTokenAmount: (0, numbers_1.parseValue)("1000", USDC.decimals), // 1000 USDC
            };
            const result = (0, orderTransactions_1.buildIncreaseOrderPayload)(params);
            (0, vitest_1.expect)(result).toEqual({
                orderPayload: {
                    addresses: {
                        receiver: RECEIVER,
                        cancellationReceiver: viem_1.zeroAddress,
                        callbackContract: viem_1.zeroAddress,
                        uiFeeReceiver: UI_FEE_RECEIVER,
                        market: params.marketAddress,
                        initialCollateralToken: USDC.address,
                        swapPath: [ETH_MARKET.marketTokenAddress],
                    },
                    numbers: {
                        sizeDeltaUsd: (0, numbers_1.parseValue)("1000", numbers_1.USD_DECIMALS),
                        initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1000", USDC.decimals),
                        triggerPrice: 0n,
                        acceptablePrice: (0, numbers_1.parseValue)("1206", numbers_1.USD_DECIMALS - WETH.decimals),
                        executionFee: EXECUTION_FEE_AMOUNT,
                        callbackGasLimit: 0n,
                        minOutputAmount: 0n,
                        validFromTime: 0n,
                    },
                    orderType: orders_1.OrderType.MarketIncrease,
                    decreasePositionSwapType: orders_1.DecreasePositionSwapType.NoSwap,
                    isLong: true,
                    shouldUnwrapNativeToken: false,
                    autoCancel: false,
                    referralCode: REFERRAL_CODE,
                    dataList: [],
                },
                params,
                tokenTransfersParams: {
                    isNativePayment: false,
                    isNativeReceive: false,
                    initialCollateralTokenAddress: USDC.address,
                    initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1000", USDC.decimals),
                    tokenTransfers: [
                        {
                            amount: EXECUTION_FEE_AMOUNT,
                            destination: ORDER_VAULT_ADDRESS,
                            tokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                        },
                        {
                            amount: (0, numbers_1.parseValue)("1000", USDC.decimals),
                            destination: ORDER_VAULT_ADDRESS,
                            tokenAddress: USDC.address,
                        },
                    ],
                    payTokenAddress: USDC.address,
                    payTokenAmount: (0, numbers_1.parseValue)("1000", USDC.decimals),
                    minOutputAmount: 0n,
                    swapPath: [ETH_MARKET.marketTokenAddress],
                    value: EXECUTION_FEE_AMOUNT,
                    externalCalls: undefined,
                },
            });
        });
        (0, vitest_1.it)("Market Increase with Internal Swap", () => {
            const params = {
                ...commonParams,
                payTokenAddress: USDC.address,
                payTokenAmount: (0, numbers_1.parseValue)("1000", USDC.decimals), // 1000 USDC
                collateralTokenAddress: WETH.address,
                collateralDeltaAmount: (0, numbers_1.parseValue)("0.5", WETH.decimals), // 0.5 ETH
            };
            const result = (0, orderTransactions_1.buildIncreaseOrderPayload)(params);
            (0, vitest_1.expect)(result).toEqual({
                orderPayload: {
                    addresses: {
                        receiver: RECEIVER,
                        cancellationReceiver: viem_1.zeroAddress,
                        callbackContract: viem_1.zeroAddress,
                        uiFeeReceiver: UI_FEE_RECEIVER,
                        market: params.marketAddress,
                        initialCollateralToken: USDC.address,
                        swapPath: [ETH_MARKET.marketTokenAddress],
                    },
                    numbers: {
                        sizeDeltaUsd: (0, numbers_1.parseValue)("1000", numbers_1.USD_DECIMALS),
                        initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1000", USDC.decimals),
                        triggerPrice: 0n,
                        acceptablePrice: (0, numbers_1.parseValue)("1206", numbers_1.USD_DECIMALS - WETH.decimals),
                        executionFee: EXECUTION_FEE_AMOUNT,
                        callbackGasLimit: 0n,
                        minOutputAmount: 0n,
                        validFromTime: 0n,
                    },
                    orderType: orders_1.OrderType.MarketIncrease,
                    decreasePositionSwapType: orders_1.DecreasePositionSwapType.NoSwap,
                    isLong: true,
                    shouldUnwrapNativeToken: false,
                    autoCancel: false,
                    referralCode: REFERRAL_CODE,
                    dataList: [],
                },
                params,
                tokenTransfersParams: {
                    isNativePayment: false,
                    isNativeReceive: false,
                    initialCollateralTokenAddress: USDC.address,
                    initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1000", USDC.decimals),
                    tokenTransfers: [
                        {
                            amount: EXECUTION_FEE_AMOUNT,
                            destination: ORDER_VAULT_ADDRESS,
                            tokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                        },
                        {
                            amount: (0, numbers_1.parseValue)("1000", USDC.decimals),
                            destination: ORDER_VAULT_ADDRESS,
                            tokenAddress: USDC.address,
                        },
                    ],
                    payTokenAddress: USDC.address,
                    payTokenAmount: (0, numbers_1.parseValue)("1000", USDC.decimals),
                    minOutputAmount: 0n,
                    swapPath: [ETH_MARKET.marketTokenAddress],
                    value: EXECUTION_FEE_AMOUNT,
                    externalCalls: undefined,
                },
            });
        });
        (0, vitest_1.it)("Market Increase with External Swap, Pay With Native Token", () => {
            const params = {
                ...commonParams,
                swapPath: [],
                payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                payTokenAmount: (0, numbers_1.parseValue)("1", WETH.decimals), // 1 ETH
                collateralTokenAddress: USDC.address,
                collateralDeltaAmount: (0, numbers_1.parseValue)("1000", USDC.decimals), // 1000 USDC
                externalSwapQuote: (0, mock_1.mockExternalSwap)({
                    inToken: WETH,
                    outToken: USDC,
                    amountIn: (0, numbers_1.parseValue)("1", WETH.decimals),
                    amountOut: (0, numbers_1.parseValue)("1000", USDC.decimals),
                    priceIn: (0, numbers_1.parseValue)("1000", numbers_1.USD_DECIMALS), // $1000 per ETH
                    priceOut: (0, numbers_1.parseValue)("1", numbers_1.USD_DECIMALS), // $1 per USDC
                }),
            };
            const result = (0, orderTransactions_1.buildIncreaseOrderPayload)(params);
            (0, vitest_1.expect)(result).toEqual({
                orderPayload: {
                    addresses: {
                        receiver: RECEIVER,
                        cancellationReceiver: viem_1.zeroAddress,
                        callbackContract: viem_1.zeroAddress,
                        uiFeeReceiver: UI_FEE_RECEIVER,
                        market: params.marketAddress,
                        initialCollateralToken: USDC.address,
                        swapPath: [],
                    },
                    numbers: {
                        sizeDeltaUsd: (0, numbers_1.parseValue)("1000", numbers_1.USD_DECIMALS),
                        initialCollateralDeltaAmount: 0n,
                        triggerPrice: 0n,
                        acceptablePrice: (0, numbers_1.parseValue)("1206", numbers_1.USD_DECIMALS - WETH.decimals),
                        executionFee: EXECUTION_FEE_AMOUNT,
                        callbackGasLimit: 0n,
                        minOutputAmount: 0n,
                        validFromTime: 0n,
                    },
                    orderType: orders_1.OrderType.MarketIncrease,
                    decreasePositionSwapType: orders_1.DecreasePositionSwapType.NoSwap,
                    isLong: true,
                    shouldUnwrapNativeToken: true,
                    autoCancel: false,
                    referralCode: REFERRAL_CODE,
                    dataList: [],
                },
                params,
                tokenTransfersParams: {
                    isNativePayment: true,
                    isNativeReceive: false,
                    initialCollateralTokenAddress: USDC.address,
                    initialCollateralDeltaAmount: 0n,
                    tokenTransfers: [
                        {
                            amount: EXECUTION_FEE_AMOUNT,
                            destination: ORDER_VAULT_ADDRESS,
                            tokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                        },
                        {
                            amount: (0, numbers_1.parseValue)("1", WETH.decimals),
                            destination: EXTERNAL_HANDLER_ADDRESS,
                            tokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                        },
                    ],
                    payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                    payTokenAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                    minOutputAmount: 0n,
                    swapPath: [],
                    value: EXECUTION_FEE_AMOUNT + (0, numbers_1.parseValue)("1", WETH.decimals),
                    externalCalls: {
                        externalCallDataList: ["0x1"],
                        externalCallTargets: ["0x6352a56caadC4F1E25CD6c75970Fa768A3304e64"],
                        refundReceivers: [RECEIVER, RECEIVER],
                        refundTokens: [WETH.address, USDC.address],
                        sendAmounts: [(0, numbers_1.parseValue)("1", WETH.decimals)],
                        sendTokens: [WETH.address],
                    },
                },
            });
        });
        (0, vitest_1.it)("Market Increase with External Swap, Pay With ERC20", () => {
            const params = {
                ...commonParams,
                swapPath: [],
                payTokenAddress: USDC.address,
                payTokenAmount: (0, numbers_1.parseValue)("1000", USDC.decimals), // 1000 USDC
                collateralTokenAddress: WETH.address,
                collateralDeltaAmount: (0, numbers_1.parseValue)("0.5", WETH.decimals), // 0.5 ETH
                externalSwapQuote: (0, mock_1.mockExternalSwap)({
                    inToken: USDC,
                    outToken: WETH,
                    amountIn: (0, numbers_1.parseValue)("1000", USDC.decimals),
                    amountOut: (0, numbers_1.parseValue)("0.5", WETH.decimals),
                    priceIn: (0, numbers_1.parseValue)("1", numbers_1.USD_DECIMALS), // $1 per USDC
                    priceOut: (0, numbers_1.parseValue)("2000", numbers_1.USD_DECIMALS), // $2000 per ETH
                }),
            };
            const result = (0, orderTransactions_1.buildIncreaseOrderPayload)(params);
            (0, vitest_1.expect)(result).toEqual({
                orderPayload: {
                    addresses: {
                        receiver: RECEIVER,
                        cancellationReceiver: viem_1.zeroAddress,
                        callbackContract: viem_1.zeroAddress,
                        uiFeeReceiver: UI_FEE_RECEIVER,
                        market: params.marketAddress,
                        initialCollateralToken: WETH.address,
                        swapPath: [],
                    },
                    numbers: {
                        sizeDeltaUsd: (0, numbers_1.parseValue)("1000", numbers_1.USD_DECIMALS),
                        initialCollateralDeltaAmount: 0n,
                        triggerPrice: 0n,
                        acceptablePrice: (0, numbers_1.parseValue)("1206", numbers_1.USD_DECIMALS - WETH.decimals),
                        executionFee: EXECUTION_FEE_AMOUNT,
                        callbackGasLimit: 0n,
                        minOutputAmount: 0n,
                        validFromTime: 0n,
                    },
                    orderType: orders_1.OrderType.MarketIncrease,
                    decreasePositionSwapType: orders_1.DecreasePositionSwapType.NoSwap,
                    isLong: true,
                    shouldUnwrapNativeToken: false,
                    autoCancel: false,
                    referralCode: REFERRAL_CODE,
                    dataList: [],
                },
                params,
                tokenTransfersParams: {
                    isNativePayment: false,
                    isNativeReceive: false,
                    initialCollateralTokenAddress: WETH.address,
                    initialCollateralDeltaAmount: 0n,
                    tokenTransfers: [
                        {
                            amount: EXECUTION_FEE_AMOUNT,
                            destination: ORDER_VAULT_ADDRESS,
                            tokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                        },
                        {
                            amount: (0, numbers_1.parseValue)("1000", USDC.decimals),
                            destination: EXTERNAL_HANDLER_ADDRESS,
                            tokenAddress: USDC.address,
                        },
                    ],
                    payTokenAddress: USDC.address,
                    payTokenAmount: (0, numbers_1.parseValue)("1000", USDC.decimals),
                    minOutputAmount: 0n,
                    swapPath: [],
                    value: EXECUTION_FEE_AMOUNT,
                    externalCalls: {
                        externalCallDataList: ["0x1"],
                        externalCallTargets: ["0x6352a56caadC4F1E25CD6c75970Fa768A3304e64"],
                        refundReceivers: [RECEIVER, RECEIVER],
                        refundTokens: [USDC.address, WETH.address],
                        sendAmounts: [(0, numbers_1.parseValue)("1000", USDC.decimals)],
                        sendTokens: [USDC.address],
                    },
                },
            });
        });
        (0, vitest_1.it)("Limit Increase", () => {
            const params = {
                ...commonParams,
                payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                payTokenAmount: (0, numbers_1.parseValue)("1", WETH.decimals), // 1 ETH
                orderType: orders_1.OrderType.LimitIncrease,
                triggerPrice: (0, numbers_1.parseValue)("1200", numbers_1.USD_DECIMALS), // $1200
            };
            const result = (0, orderTransactions_1.buildIncreaseOrderPayload)(params);
            (0, vitest_1.expect)(result).toEqual({
                orderPayload: {
                    addresses: {
                        receiver: RECEIVER,
                        cancellationReceiver: viem_1.zeroAddress,
                        callbackContract: viem_1.zeroAddress,
                        uiFeeReceiver: UI_FEE_RECEIVER,
                        market: params.marketAddress,
                        initialCollateralToken: WETH.address,
                        swapPath: [ETH_MARKET.marketTokenAddress],
                    },
                    numbers: {
                        sizeDeltaUsd: (0, numbers_1.parseValue)("1000", numbers_1.USD_DECIMALS),
                        initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                        triggerPrice: (0, numbers_1.parseValue)("1200", numbers_1.USD_DECIMALS - WETH.decimals),
                        acceptablePrice: (0, numbers_1.parseValue)("1206", numbers_1.USD_DECIMALS - WETH.decimals),
                        executionFee: EXECUTION_FEE_AMOUNT,
                        callbackGasLimit: 0n,
                        minOutputAmount: 0n,
                        validFromTime: 0n,
                    },
                    orderType: orders_1.OrderType.LimitIncrease,
                    decreasePositionSwapType: orders_1.DecreasePositionSwapType.NoSwap,
                    isLong: true,
                    shouldUnwrapNativeToken: true,
                    autoCancel: false,
                    referralCode: REFERRAL_CODE,
                    dataList: [],
                },
                params,
                tokenTransfersParams: {
                    isNativePayment: true,
                    isNativeReceive: false,
                    initialCollateralTokenAddress: WETH.address,
                    initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                    tokenTransfers: [
                        {
                            amount: EXECUTION_FEE_AMOUNT + (0, numbers_1.parseValue)("1", WETH.decimals),
                            destination: ORDER_VAULT_ADDRESS,
                            tokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                        },
                    ],
                    payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                    payTokenAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                    minOutputAmount: 0n,
                    swapPath: [ETH_MARKET.marketTokenAddress],
                    value: EXECUTION_FEE_AMOUNT + (0, numbers_1.parseValue)("1", WETH.decimals),
                    externalCalls: undefined,
                },
            });
        });
        (0, vitest_1.it)("TWAP Increase Long Pay with Native Token", () => {
            const params = {
                ...commonParams,
                payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                payTokenAmount: (0, numbers_1.parseValue)("1", WETH.decimals), // 1 ETH
                orderType: orders_1.OrderType.MarketIncrease,
            };
            const twapParams = {
                duration: {
                    hours: 1,
                    minutes: 0,
                },
                numberOfParts: 4,
            };
            const result = (0, orderTransactions_1.buildTwapOrdersPayloads)(params, twapParams);
            // Calculate time intervals for each part
            const totalDurationInSeconds = twapParams.duration.hours * 3600 + twapParams.duration.minutes * 60;
            const startTime = Math.ceil(Date.now() / 1000);
            const intervalInSeconds = totalDurationInSeconds / (twapParams.numberOfParts - 1);
            const expectedOrders = Array.from({ length: twapParams.numberOfParts }, (_, i) => {
                const validFromTime = BigInt(Math.floor(startTime + intervalInSeconds * i));
                const uiFeeReceiver = `0xff00000000000000000000000000000004800001`;
                const decoded = (0, uiFeeReceiver_1.decodeTwapUiFeeReceiver)(uiFeeReceiver);
                (0, vitest_1.expect)(decoded).toEqual({
                    isExpress: false,
                    twapId: "8000",
                    numberOfParts: twapParams.numberOfParts,
                });
                return {
                    orderPayload: {
                        addresses: {
                            receiver: RECEIVER,
                            cancellationReceiver: viem_1.zeroAddress,
                            callbackContract: viem_1.zeroAddress,
                            uiFeeReceiver,
                            market: params.marketAddress,
                            initialCollateralToken: WETH.address,
                            swapPath: [ETH_MARKET.marketTokenAddress],
                        },
                        numbers: {
                            sizeDeltaUsd: (0, numbers_1.parseValue)("250", numbers_1.USD_DECIMALS), // 1000/4
                            initialCollateralDeltaAmount: params.payTokenAmount / 4n, // 1/4
                            triggerPrice: numbers_1.MaxUint256,
                            acceptablePrice: numbers_1.MaxUint256,
                            executionFee: EXECUTION_FEE_AMOUNT / 4n,
                            callbackGasLimit: 0n,
                            minOutputAmount: 0n,
                            validFromTime,
                        },
                        orderType: orders_1.OrderType.LimitIncrease,
                        decreasePositionSwapType: orders_1.DecreasePositionSwapType.NoSwap,
                        isLong: true,
                        shouldUnwrapNativeToken: true,
                        autoCancel: false,
                        referralCode: REFERRAL_CODE,
                        dataList: [],
                    },
                    params: {
                        ...params,
                        acceptablePrice: numbers_1.MaxUint256,
                        triggerPrice: numbers_1.MaxUint256,
                        payTokenAmount: params.payTokenAmount / 4n,
                        collateralDeltaAmount: params.collateralDeltaAmount / 4n,
                        executionFeeAmount: params.executionFeeAmount / 4n,
                        sizeDeltaUsd: params.sizeDeltaUsd / 4n,
                        sizeDeltaInTokens: params.sizeDeltaInTokens / 4n,
                        orderType: orders_1.OrderType.LimitIncrease,
                        allowedSlippage: 0,
                        uiFeeReceiver,
                        validFromTime,
                    },
                    tokenTransfersParams: {
                        isNativePayment: true,
                        isNativeReceive: false,
                        initialCollateralTokenAddress: WETH.address,
                        initialCollateralDeltaAmount: (0, numbers_1.parseValue)("0.25", WETH.decimals),
                        tokenTransfers: [
                            {
                                amount: EXECUTION_FEE_AMOUNT / 4n + (0, numbers_1.parseValue)("0.25", WETH.decimals),
                                destination: ORDER_VAULT_ADDRESS,
                                tokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                            },
                        ],
                        payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                        payTokenAmount: (0, numbers_1.parseValue)("0.25", WETH.decimals),
                        minOutputAmount: 0n,
                        swapPath: [ETH_MARKET.marketTokenAddress],
                        value: EXECUTION_FEE_AMOUNT / 4n + (0, numbers_1.parseValue)("0.25", WETH.decimals),
                        externalCalls: undefined,
                    },
                };
            });
            (0, vitest_1.expect)(result).toEqual(expectedOrders);
            (0, vitest_1.expect)(expectedOrders.every((co) => (0, orderTransactions_1.getIsTwapOrderPayload)(co.orderPayload))).toBe(true);
        });
        (0, vitest_1.it)("TWAP Increase Short Pay with Native Token", () => {
            const params = {
                ...commonParams,
                payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                payTokenAmount: (0, numbers_1.parseValue)("1", WETH.decimals), // 1 ETH
                orderType: orders_1.OrderType.MarketIncrease,
                isLong: false,
            };
            const twapParams = {
                duration: {
                    hours: 1,
                    minutes: 0,
                },
                numberOfParts: 4,
            };
            const result = (0, orderTransactions_1.buildTwapOrdersPayloads)(params, twapParams);
            // Calculate time intervals for each part
            const totalDurationInSeconds = twapParams.duration.hours * 3600 + twapParams.duration.minutes * 60;
            const startTime = Math.ceil(Date.now() / 1000);
            const intervalInSeconds = totalDurationInSeconds / (twapParams.numberOfParts - 1);
            const expectedOrders = Array.from({ length: twapParams.numberOfParts }, (_, i) => {
                const validFromTime = BigInt(Math.floor(startTime + intervalInSeconds * i));
                const uiFeeReceiver = `0xff00000000000000000000000000000004800001`;
                const decoded = (0, uiFeeReceiver_1.decodeTwapUiFeeReceiver)(uiFeeReceiver);
                (0, vitest_1.expect)(decoded).toEqual({
                    isExpress: false,
                    twapId: "8000",
                    numberOfParts: twapParams.numberOfParts,
                });
                return {
                    orderPayload: {
                        addresses: {
                            receiver: RECEIVER,
                            cancellationReceiver: viem_1.zeroAddress,
                            callbackContract: viem_1.zeroAddress,
                            uiFeeReceiver,
                            market: params.marketAddress,
                            initialCollateralToken: WETH.address,
                            swapPath: [ETH_MARKET.marketTokenAddress],
                        },
                        numbers: {
                            sizeDeltaUsd: (0, numbers_1.parseValue)("250", numbers_1.USD_DECIMALS), // 1000/4
                            initialCollateralDeltaAmount: params.payTokenAmount / 4n, // 1/4
                            triggerPrice: 0n,
                            acceptablePrice: 0n,
                            executionFee: EXECUTION_FEE_AMOUNT / 4n,
                            callbackGasLimit: 0n,
                            minOutputAmount: 0n,
                            validFromTime,
                        },
                        orderType: orders_1.OrderType.LimitIncrease,
                        decreasePositionSwapType: orders_1.DecreasePositionSwapType.NoSwap,
                        isLong: false,
                        shouldUnwrapNativeToken: true,
                        autoCancel: false,
                        referralCode: REFERRAL_CODE,
                        dataList: [],
                    },
                    params: {
                        ...params,
                        acceptablePrice: 0n,
                        triggerPrice: 0n,
                        payTokenAmount: params.payTokenAmount / 4n,
                        collateralDeltaAmount: params.collateralDeltaAmount / 4n,
                        executionFeeAmount: params.executionFeeAmount / 4n,
                        sizeDeltaUsd: params.sizeDeltaUsd / 4n,
                        sizeDeltaInTokens: params.sizeDeltaInTokens / 4n,
                        orderType: orders_1.OrderType.LimitIncrease,
                        allowedSlippage: 0,
                        uiFeeReceiver,
                        validFromTime,
                    },
                    tokenTransfersParams: {
                        isNativePayment: true,
                        isNativeReceive: false,
                        initialCollateralTokenAddress: WETH.address,
                        initialCollateralDeltaAmount: params.payTokenAmount / 4n,
                        tokenTransfers: [
                            {
                                amount: EXECUTION_FEE_AMOUNT / 4n + params.payTokenAmount / 4n,
                                destination: ORDER_VAULT_ADDRESS,
                                tokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                            },
                        ],
                        payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                        payTokenAmount: params.payTokenAmount / 4n,
                        minOutputAmount: 0n,
                        swapPath: [ETH_MARKET.marketTokenAddress],
                        value: EXECUTION_FEE_AMOUNT / 4n + params.payTokenAmount / 4n,
                        externalCalls: undefined,
                    },
                };
            });
            (0, vitest_1.expect)(result).toEqual(expectedOrders);
            (0, vitest_1.expect)(expectedOrders.every((co) => (0, orderTransactions_1.getIsTwapOrderPayload)(co.orderPayload))).toBe(true);
        });
        (0, vitest_1.it)("Market Increase with Undefined Optional Parameters", () => {
            const params = {
                ...commonParams,
                payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                payTokenAmount: (0, numbers_1.parseValue)("1", WETH.decimals), // 1 ETH
                orderType: orders_1.OrderType.MarketIncrease,
                uiFeeReceiver: undefined,
                referralCode: undefined,
                validFromTime: undefined,
            };
            const result = (0, orderTransactions_1.buildIncreaseOrderPayload)(params);
            (0, vitest_1.expect)(result).toEqual({
                orderPayload: {
                    addresses: {
                        receiver: RECEIVER,
                        cancellationReceiver: viem_1.zeroAddress,
                        callbackContract: viem_1.zeroAddress,
                        uiFeeReceiver: viem_1.zeroAddress,
                        market: params.marketAddress,
                        initialCollateralToken: WETH.address,
                        swapPath: [ETH_MARKET.marketTokenAddress],
                    },
                    numbers: {
                        sizeDeltaUsd: (0, numbers_1.parseValue)("1000", numbers_1.USD_DECIMALS),
                        initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                        triggerPrice: 0n,
                        acceptablePrice: (0, numbers_1.parseValue)("1206", numbers_1.USD_DECIMALS - WETH.decimals),
                        executionFee: EXECUTION_FEE_AMOUNT,
                        callbackGasLimit: 0n,
                        minOutputAmount: 0n,
                        validFromTime: 0n,
                    },
                    orderType: orders_1.OrderType.MarketIncrease,
                    decreasePositionSwapType: orders_1.DecreasePositionSwapType.NoSwap,
                    isLong: true,
                    shouldUnwrapNativeToken: true,
                    autoCancel: false,
                    referralCode: viem_1.zeroHash,
                    dataList: [],
                },
                params,
                tokenTransfersParams: {
                    isNativePayment: true,
                    isNativeReceive: false,
                    initialCollateralTokenAddress: WETH.address,
                    initialCollateralDeltaAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                    tokenTransfers: [
                        {
                            amount: EXECUTION_FEE_AMOUNT + (0, numbers_1.parseValue)("1", WETH.decimals),
                            destination: ORDER_VAULT_ADDRESS,
                            tokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                        },
                    ],
                    payTokenAddress: tokens_1.NATIVE_TOKEN_ADDRESS,
                    payTokenAmount: (0, numbers_1.parseValue)("1", WETH.decimals),
                    minOutputAmount: 0n,
                    swapPath: [ETH_MARKET.marketTokenAddress],
                    value: EXECUTION_FEE_AMOUNT + (0, numbers_1.parseValue)("1", WETH.decimals),
                    externalCalls: undefined,
                },
            });
        });
    });
});
